<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 13 MC3 somatic variant calling pipeline | Bioinformatics tools and pipelines with R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and pipeplines with Common Workflow Language using the Rcwl package.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 13 MC3 somatic variant calling pipeline | Bioinformatics tools and pipelines with R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and pipeplines with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 MC3 somatic variant calling pipeline | Bioinformatics tools and pipelines with R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and pipeplines with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu">


<meta name="date" content="2019-03-13">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="summarize-qc.html">
<link rel="next" href="references.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/d3-3.3.8/d3.min.js"></script>
<script src="libs/dagre-0.4.0/dagre-d3.min.js"></script>
<link href="libs/mermaid-0.3.0/dist/mermaid.css" rel="stylesheet" />
<script src="libs/mermaid-0.3.0/dist/mermaid.slim.min.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/chromatography-0.1/chromatography.js"></script>
<script src="libs/DiagrammeR-binding-1.0.0/DiagrammeR.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>1.1</b> R package installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>1.2</b> System requirements</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>1.3</b> Docker</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>1.4</b> Structure of the book</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>1.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#common-workflow-language"><i class="fa fa-check"></i><b>2.1</b> Common Workflow Language</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#first-example"><i class="fa fa-check"></i><b>2.2</b> First example</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#test-run"><i class="fa fa-check"></i><b>2.3</b> Test run</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="input-parameters.html"><a href="input-parameters.html"><i class="fa fa-check"></i><b>3</b> Input Parameters</a><ul>
<li class="chapter" data-level="3.1" data-path="input-parameters.html"><a href="input-parameters.html#essential-input-parameters"><i class="fa fa-check"></i><b>3.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="3.2" data-path="input-parameters.html"><a href="input-parameters.html#array-inputs"><i class="fa fa-check"></i><b>3.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="output-parameters.html"><a href="output-parameters.html"><i class="fa fa-check"></i><b>4</b> Output Parameters</a><ul>
<li class="chapter" data-level="4.1" data-path="output-parameters.html"><a href="output-parameters.html#capturing-output"><i class="fa fa-check"></i><b>4.1</b> Capturing Output</a></li>
<li class="chapter" data-level="4.2" data-path="output-parameters.html"><a href="output-parameters.html#array-outputs"><i class="fa fa-check"></i><b>4.2</b> Array Outputs</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>5</b> Writing Pipeline</a><ul>
<li class="chapter" data-level="5.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>5.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="5.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>5.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="running-tools-in-docker.html"><a href="running-tools-in-docker.html"><i class="fa fa-check"></i><b>6</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="7" data-path="running-tools-in-cluster-server.html"><a href="running-tools-in-cluster-server.html"><i class="fa fa-check"></i><b>7</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="8" data-path="web-application.html"><a href="web-application.html"><i class="fa fa-check"></i><b>8</b> Web Application</a><ul>
<li class="chapter" data-level="8.1" data-path="web-application.html"><a href="web-application.html#cwlparam-example"><i class="fa fa-check"></i><b>8.1</b> cwlParam example</a></li>
<li class="chapter" data-level="8.2" data-path="web-application.html"><a href="web-application.html#cwlparam-to-shiny-app"><i class="fa fa-check"></i><b>8.2</b> cwlParam to Shiny App</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html"><i class="fa fa-check"></i><b>9</b> RcwlPipelines tools</a><ul>
<li class="chapter" data-level="9.1" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html#rcwl-scripts"><i class="fa fa-check"></i><b>9.1</b> Rcwl scripts</a></li>
<li class="chapter" data-level="9.2" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html#build-a-pipeline"><i class="fa fa-check"></i><b>9.2</b> Build a pipeline</a></li>
<li class="chapter" data-level="9.3" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html#rcwlpipelines-summary"><i class="fa fa-check"></i><b>9.3</b> RcwlPipelines summary</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html"><i class="fa fa-check"></i><b>10</b> DNASeq alignment pipeline</a><ul>
<li class="chapter" data-level="10.1" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html#summary"><i class="fa fa-check"></i><b>10.1</b> Summary</a></li>
<li class="chapter" data-level="10.2" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html#prepara-data"><i class="fa fa-check"></i><b>10.2</b> Prepara data</a></li>
<li class="chapter" data-level="10.3" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html#run-in-cluster"><i class="fa fa-check"></i><b>10.3</b> Run in cluster</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="rnaseq-pipeline.html"><a href="rnaseq-pipeline.html"><i class="fa fa-check"></i><b>11</b> RNASeq pipeline</a><ul>
<li class="chapter" data-level="11.1" data-path="rnaseq-pipeline.html"><a href="rnaseq-pipeline.html#prepare-data"><i class="fa fa-check"></i><b>11.1</b> Prepare data</a></li>
<li class="chapter" data-level="11.2" data-path="rnaseq-pipeline.html"><a href="rnaseq-pipeline.html#submit-pipeline-with-sge"><i class="fa fa-check"></i><b>11.2</b> Submit pipeline with SGE</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="summarize-qc.html"><a href="summarize-qc.html"><i class="fa fa-check"></i><b>12</b> Summarize QC</a></li>
<li class="chapter" data-level="13" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html"><i class="fa fa-check"></i><b>13</b> MC3 somatic variant calling pipeline</a><ul>
<li class="chapter" data-level="13.1" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#prepare-data-1"><i class="fa fa-check"></i><b>13.1</b> Prepare data</a></li>
<li class="chapter" data-level="13.2" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#run-mc3-pipeline"><i class="fa fa-check"></i><b>13.2</b> Run MC3 pipeline</a></li>
<li class="chapter" data-level="13.3" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#gatk4-germline-variant-calling-pipeline"><i class="fa fa-check"></i><b>13.3</b> GATK4 germline variant calling pipeline</a></li>
<li class="chapter" data-level="13.4" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#gatk-alignment"><i class="fa fa-check"></i><b>13.4</b> GATK Alignment</a></li>
<li class="chapter" data-level="13.5" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#haplotypecaller"><i class="fa fa-check"></i><b>13.5</b> HaplotypeCaller</a></li>
<li class="chapter" data-level="13.6" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#joint-discovery"><i class="fa fa-check"></i><b>13.6</b> Joint Discovery</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines with R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mc3-somatic-variant-calling-pipeline" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> MC3 somatic variant calling pipeline</h1>
<p>The Multi-Center Mutation Calling in Multiple Cancers project (MC3) pipeline was developed by TCGA to generate a comprehensive encyclopedia of somatic mutation calls by applying an ensemble of seven mutation-calling algorithms with scoring and artifact filtering. More details can be found in this paper:
<a href="https://www.cell.com/fulltext/S2405-4712(18)30096-6">Scalable Open Science Approach for Mutation Calling of Tumor Exomes Using Multiple Genomic Pipelines</a></p>
<p>The <code>mc3</code> pipeline is available at <a href="https://github.com/OpenGenomics/mc3" class="uri">https://github.com/OpenGenomics/mc3</a>. All required software have been deployed in cloud with docker.</p>
<p>The pipeline has been imported and contained in the <code>RcwlPipelines</code> pacakge, which contains two major steps (markID step was removed):</p>
<ul>
<li>Call variants by 7 pipelines</li>
<li>Merge VCF and convert to MAF</li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(mc3)
<span class="kw">short</span>(mc3)</code></pre>
<pre><code>## inputs:
## - tumorID
## - normalID
## - tumor
## - normal
## - bed_file
## - centromere
## - cosmic
## - dbsnp
## - refFasta
## - vepData
## outputs:
## - outmaf
## - outvcf
## steps:
## - call_variants
## - convert</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotCWL</span>(mc3)</code></pre>
<div id="htmlwidget-b699b1de6ebcdeb0d8e0" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-b699b1de6ebcdeb0d8e0">{"x":{"diagram":"graph TB\nC(tumor)-->M{call_variants}\nD(normal)-->M{call_variants}\nF(centromere)-->M{call_variants}\nE(bed_file)-->M{call_variants}\nG(cosmic)-->M{call_variants}\nH(dbsnp)-->M{call_variants}\nI(refFasta)-->M{call_variants}\nA(tumorID)-->M{call_variants}\nB(normalID)-->M{call_variants}\nB(normalID)-->N{convert}\nA(tumorID)-->N{convert}\nM{call_variants}-->|museVCF|N{convert}\nM{call_variants}-->|mutectVCF|N{convert}\nM{call_variants}-->|somaticsniperVCF|N{convert}\nM{call_variants}-->|varscansVCF|N{convert}\nM{call_variants}-->|varscaniVCF|N{convert}\nM{call_variants}-->|radiaVCF|N{convert}\nM{call_variants}-->|pindelVCF|N{convert}\nM{call_variants}-->|indelocatorVCF|N{convert}\nI(refFasta)-->N{convert}\nJ(vepData)-->N{convert}\nC(tumor)-->N{convert}\nD(normal)-->N{convert}\nN{convert}-->K((outmaf))\nN{convert}-->L((outvcf))"},"evals":[],"jsHooks":[]}</script>
<p>Two steps are included.<br />
1. <code>call_variants</code>: To call variants by 7 pipelines.</p>
<pre class="sourceCode r"><code class="sourceCode r">callVar &lt;-<span class="st"> </span><span class="kw">readCWL</span>(<span class="kw">runs</span>(mc3)<span class="op">$</span>call_variants)
<span class="kw">plotCWL</span>(callVar)</code></pre>
<div id="htmlwidget-3fc637841f58bb456c23" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-3fc637841f58bb456c23">{"x":{"diagram":"graph TB\nC(reference)-->S{prep_ref}\nB(normal)-->T{normal_pileup}\nS{prep_ref}-->|prepped_ref|T{normal_pileup}\nA(tumor)-->U{tumor_pileup}\nS{prep_ref}-->|prepped_ref|U{tumor_pileup}\nU{tumor_pileup}-->|output|V{varscan}\nT{normal_pileup}-->|output|V{varscan}\nV{varscan}-->N((varscaniVCF))\nA(tumor)-->W{muse}\nB(normal)-->W{muse}\nS{prep_ref}-->|prepped_ref|W{muse}\nD(dbsnp)-->W{muse}\nH(number_of_procs)-->W{muse}\nA(tumor)-->X{mutect}\nB(normal)-->X{mutect}\nS{prep_ref}-->|prepped_ref|X{mutect}\nE(cosmic)-->X{mutect}\nD(dbsnp)-->X{mutect}\nH(number_of_procs)-->X{mutect}\nX{mutect}-->P((mutectVCF))\nA(tumor)-->Y{radia}\nB(normal)-->Y{radia}\nS{prep_ref}-->|prepped_ref|Y{radia}\nH(number_of_procs)-->Y{radia}\nA(tumor)-->Z{somaticsniper}\nB(normal)-->Z{somaticsniper}\nS{prep_ref}-->|prepped_ref|Z{somaticsniper}\nA(tumor)-->A1{pindel}\nB(normal)-->A1{pindel}\nS{prep_ref}-->|prepped_ref|A1{pindel}\nF(centromere)-->A1{pindel}\nH(number_of_procs)-->A1{pindel}\nA(tumor)-->B2{indelocator}\nB(normal)-->B2{indelocator}\nS{prep_ref}-->|prepped_ref|B2{indelocator}\nG(bed_file)-->B2{indelocator}\nB2{indelocator}-->|mutations|C3{reformat_indelocator}\nY{radia}-->|mutations|D4{fpfilter}\nV{varscan}-->|snp_vcf|D4{fpfilter}\nZ{somaticsniper}-->|mutations|D4{fpfilter}\nW{muse}-->|mutations|D4{fpfilter}\nA1{pindel}-->|somatic_vcf|D4{fpfilter}\nC3{reformat_indelocator}-->|mutations|D4{fpfilter}\nA(tumor)-->D4{fpfilter}\nB(normal)-->D4{fpfilter}\nS{prep_ref}-->|prepped_ref|D4{fpfilter}\nI(tumorID)-->D4{fpfilter}\nJ(normalID)-->D4{fpfilter}\nD4{fpfilter}-->Q((radiaVCF))\nD4{fpfilter}-->L((somaticsniperVCF))\nD4{fpfilter}-->M((varscansVCF))\nD4{fpfilter}-->O((museVCF))\nD4{fpfilter}-->K((pindelVCF))\nD4{fpfilter}-->R((indelocatorVCF))"},"evals":[],"jsHooks":[]}</script>
<ol start="2" style="list-style-type: decimal">
<li><code>covert</code>: To merge VCFs and convert to MAF</li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r">conv &lt;-<span class="st"> </span><span class="kw">readCWL</span>(<span class="kw">runs</span>(mc3)<span class="op">$</span>convert)
<span class="kw">plotCWL</span>(conv)</code></pre>
<div id="htmlwidget-3a914e929387c0928a1e" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-3a914e929387c0928a1e">{"x":{"diagram":"graph TB\nC(museVCF)-->W{sort_muse}\nW{sort_muse}-->|output_vcf|X{standardize_muse}\nK(refFasta)-->X{standardize_muse}\nB(tumorID)-->X{standardize_muse}\nA(normalID)-->X{standardize_muse}\nD(mutectVCF)-->Y{sort_mutect}\nY{sort_mutect}-->|output_vcf|Z{standardize_mutect}\nK(refFasta)-->Z{standardize_mutect}\nB(tumorID)-->Z{standardize_mutect}\nA(normalID)-->Z{standardize_mutect}\nE(somaticsniperVCF)-->A1{sort_somaticsniper}\nA1{sort_somaticsniper}-->|output_vcf|B2{standardize_somaticsniper}\nK(refFasta)-->B2{standardize_somaticsniper}\nB(tumorID)-->B2{standardize_somaticsniper}\nA(normalID)-->B2{standardize_somaticsniper}\nF(varscansVCF)-->C3{sort_varscans}\nC3{sort_varscans}-->|output_vcf|D4{standardize_varscans}\nK(refFasta)-->D4{standardize_varscans}\nB(tumorID)-->D4{standardize_varscans}\nA(normalID)-->D4{standardize_varscans}\nG(varscaniVCF)-->E5{sort_varscani}\nE5{sort_varscani}-->|output_vcf|F6{standardize_varscani}\nK(refFasta)-->F6{standardize_varscani}\nB(tumorID)-->F6{standardize_varscani}\nA(normalID)-->F6{standardize_varscani}\nH(radiaVCF)-->G7{sort_radia}\nG7{sort_radia}-->|output_vcf|H8{standardize_radia}\nK(refFasta)-->H8{standardize_radia}\nB(tumorID)-->H8{standardize_radia}\nA(normalID)-->H8{standardize_radia}\nJ(indelocatorVCF)-->I9{sort_indelocator}\nI9{sort_indelocator}-->|output_vcf|J10{standardize_indelocator}\nK(refFasta)-->J10{standardize_indelocator}\nB(tumorID)-->J10{standardize_indelocator}\nA(normalID)-->J10{standardize_indelocator}\nI(pindelVCF)-->K11{sort_pindel}\nK11{sort_pindel}-->|output_vcf|L12{standardize_pindel}\nK(refFasta)-->L12{standardize_pindel}\nB(tumorID)-->L12{standardize_pindel}\nA(normalID)-->L12{standardize_pindel}\nX{standardize_muse}-->|vcf|M13{rehead_vcfs}\nZ{standardize_mutect}-->|vcf|M13{rehead_vcfs}\nB2{standardize_somaticsniper}-->|vcf|M13{rehead_vcfs}\nF6{standardize_varscani}-->|vcf|M13{rehead_vcfs}\nD4{standardize_varscans}-->|vcf|M13{rehead_vcfs}\nH8{standardize_radia}-->|vcf|M13{rehead_vcfs}\nL12{standardize_pindel}-->|vcf|M13{rehead_vcfs}\nJ10{standardize_indelocator}-->|vcf|M13{rehead_vcfs}\nM(tumor_analysis_uuid)-->M13{rehead_vcfs}\nN(tumor_bam_name)-->M13{rehead_vcfs}\nO(tumor_aliquot_uuid)-->M13{rehead_vcfs}\nB(tumorID)-->M13{rehead_vcfs}\nP(normal_analysis_uuid)-->M13{rehead_vcfs}\nQ(normal_bam_name)-->M13{rehead_vcfs}\nR(normal_aliquot_uuid)-->M13{rehead_vcfs}\nA(normalID)-->M13{rehead_vcfs}\nS(platform)-->M13{rehead_vcfs}\nT(center)-->M13{rehead_vcfs}\nM13{rehead_vcfs}-->|reheaded_muse|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_mutect|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_somsniper|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_varscans|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_varscani|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_radia|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_indelocator|N14{merge_vcf}\nM13{rehead_vcfs}-->|reheaded_pindel|N14{merge_vcf}\nN14{merge_vcf}-->|output_vcf|O15{vcf2maf}\nK(refFasta)-->O15{vcf2maf}\nL(vepData)-->O15{vcf2maf}\nB(tumorID)-->O15{vcf2maf}\nA(normalID)-->O15{vcf2maf}\nO15{vcf2maf}-->U((outmaf))\nO15{vcf2maf}-->V((vepvcf))"},"evals":[],"jsHooks":[]}</script>
<p>The merged VCF and converted MAF files will be collected to the output
folder.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">outputs</span>(mc3)</code></pre>
<pre><code>## List of length 2
## names(2): outmaf outvcf</code></pre>
<p>```</p>
<div id="prepare-data-1" class="section level2">
<h2><span class="header-section-number">13.1</span> Prepare data</h2>
<p>Testing somatic mutation data can be download from: <a href="https://github.com/genome/somatic-snv-test-data" class="uri">https://github.com/genome/somatic-snv-test-data</a>.</p>
<ol style="list-style-type: decimal">
<li>Input list <code>inputList</code>.
The tumorID/normalID must be consistent with <code>SM</code> from BAM read group.</li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r">inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">tumorID=</span><span class="kw">list</span>(<span class="dt">test=</span><span class="st">&quot;NA12892&quot;</span>),
                  <span class="dt">normalID=</span><span class="kw">list</span>(<span class="dt">test=</span><span class="st">&quot;NA12878&quot;</span>),
                  <span class="dt">tumor=</span><span class="kw">list</span>(<span class="dt">test=</span><span class="st">&quot;apps/DNASeq/data/tumor.bam&quot;</span>),
                  <span class="dt">normal=</span><span class="kw">list</span>(<span class="dt">test=</span><span class="st">&quot;apps/DNASeq/data/normal.bam&quot;</span>))</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Parameter list <code>paramList</code>.</li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r">paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">bed_file=</span><span class="st">&quot;apps/data/mc3/gaf_20111020+broad_wex_1.1_hg19.bed&quot;</span>,
                  <span class="dt">centromere=</span><span class="st">&quot;apps/data/mc3/centromere_hg19.bed&quot;</span>,
                  <span class="dt">cosmic=</span><span class="st">&quot;apps/data/mc3/hg19_cosmic_v54_120711.vcf.gz&quot;</span>,
                  <span class="dt">dbsnp=</span><span class="st">&quot;apps/data/mc3/dbsnp_134_b37.leftAligned.vcf.gz&quot;</span>,
                  <span class="dt">refFasta=</span><span class="st">&quot;apps/data/human_g1k_v37.fa.gz&quot;</span>,
                  <span class="dt">vepData=</span><span class="st">&quot;apps/data/.vep/&quot;</span>)</code></pre>
</div>
<div id="run-mc3-pipeline" class="section level2">
<h2><span class="header-section-number">13.2</span> Run MC3 pipeline</h2>
<pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(mc3, <span class="dt">outdir =</span> <span class="st">&quot;apps/DNASeq/output/mc3&quot;</span>,
                   <span class="dt">inputList =</span> inputList, <span class="dt">paramList =</span> paramList,
                   <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">1</span>, <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,
                                             <span class="dt">template =</span> <span class="st">&quot;apps/data/sge_centos7.tmpl&quot;</span>,
                                             <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">2</span>,
                                                              <span class="dt">queue =</span> <span class="st">&quot;centos7.q&quot;</span>)))</code></pre>
<p>The final VCF was filtered and merged from the outputs of different
pipelines and annotated by VEP. The converted MAF file was also
generated.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dir</span>(<span class="st">&quot;apps/DNASeq/output/mc3/test&quot;</span>)</code></pre>
<pre><code>## [1] &quot;merged.vep.vcf&quot; &quot;vep.maf&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">vcf &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;apps/DNASeq/output/mc3/test/merged.vep.vcf&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">head</span>(vcf)</code></pre>
<pre><code>##   V1       V2          V3 V4 V5  V6   V7
## 1 21 10400299           .  A  T 0.0 PASS
## 2 21 10400380           .  C  T   . PASS
## 3 21 10402435   rs2948877  G  A   . PASS
## 4 21 10402715           .  G  A 0.0 PASS
## 5 21 10402795 rs148043841  G  T   . PASS
## 6 21 10402985           .  G GA   . PASS
##                                                                                                                                                                                                V8
## 1                                               CENTERS=RADIA|VARSCANS|MUSE|SOMATICSNIPER;CSQ=T|intergenic_variant|MODIFIER|||||||||||||||rs370695467|1||||1|SNV|1|||||||||||||||||||||||||||||||
## 2                                                          CENTERS=SOMATICSNIPER|RADIA|VARSCANS|MUSE;CSQ=T|intergenic_variant|MODIFIER||||||||||||||||1||||1|SNV|1|||||||||||||||||||||||||||||||
## 3 CENTERS=MUSE|RADIA|VARSCANS|SOMATICSNIPER;CSQ=A|intergenic_variant|MODIFIER|||||||||||||||rs2948877|1||||1|SNV|1||||||||||||||||A:0.3626|A:0.4266|A:0.3228|A:0.3075|A:0.2913|A:0.4346||||||||||
## 4                                                               CENTERS=RADIA|VARSCANS|MUSE;CSQ=A|intergenic_variant|MODIFIER|||||||||||||||rs2948878|1||||1|SNV|1|||||||||||||||||||||||||||||||
## 5                                                             CENTERS=MUSE|RADIA|VARSCANS;CSQ=T|intergenic_variant|MODIFIER|||||||||||||||rs373568457|1||||1|SNV|1|||||||||||||||||||||||||||||||
## 6                                                          CENTERS=VARSCANI*|PINDEL;CSQ=A|intergenic_variant|MODIFIER|||||||||||||||rs375209288|1||||1|insertion|1|||||||||||||||||||||||||||||||
##         V9           V10            V11
## 1 GT:DP:AD 0/0:140:140,0   0/1:92:71,20
## 2 GT:DP:AD 0/0:160:160,0  0/1:117:99,18
## 3 GT:DP:AD 0/0:167:167,0  0/1:124:97,27
## 4 GT:DP:AD 0/0:145:145,0  0/1:117:97,20
## 5 GT:DP:AD 0/0:163:161,2 0/1:127:108,19
## 6 GT:DP:AD   0/0:88:88,0    0/1:82:75,7</code></pre>
</div>
<div id="gatk4-germline-variant-calling-pipeline" class="section level2">
<h2><span class="header-section-number">13.3</span> GATK4 germline variant calling pipeline</h2>
<p>The GATK 4 best practice pipeline for germline variant calling was
implemented with Workflow Description Language (WDL), which is similar
to cwl and requires
<a href="https://github.com/broadinstitute/cromwell">cromwell</a> to run the
pipelines. We wrapped the WDL pipeline into 3 steps with <code>Rcwl</code>. The
details of the pipeline can be find here:
<a href="https://software.broadinstitute.org/gatk/best-practices/workflow?id=11145" class="uri">https://software.broadinstitute.org/gatk/best-practices/workflow?id=11145</a></p>
<div class="figure">
<img src="https://us.v-cdn.net/5019796/uploads/editor/mz/tzm69d8e2spl.png" alt="Germline short variant discovery (SNPs + Indels)" />
<p class="caption">Germline short variant discovery (SNPs + Indels)</p>
</div>
<p>The germline pipelines include 4 steps in WDL, paired fastq to ubam,
GATK alignment, variant calling by HaplotypeCaller and joint
genotyping. We wrapped the GATK pipelines to 3 steps using <code>Rcwl</code> for
different numbers of computing nodes requirements. The wrapped
pipelines can help to assign inputs to the input JSON templates and
glob results from the <code>cromwell</code> outputs.</p>
<ol style="list-style-type: decimal">
<li><code>GAlign</code> GATK alignment.</li>
</ol>
<p>The fastqs, sample information and customized json files for WDL are
required as inputs. Multiple steps will run in this step, including
<code>bwa</code> alignment, mark duplicates and base quality recalibration. GATK
ready BAM files will be collected to the output directory.</p>
<ol start="2" style="list-style-type: decimal">
<li><code>hapCall</code> HaplotypeCaller.</li>
</ol>
<p>The GATK ready BAM and customized json files are inputs in this
step. The local paths of GATK bundle files are required to be modified
in your json file. A “gVCF” files will be generated.</p>
<ol start="3" style="list-style-type: decimal">
<li><code>jdCall</code> Joint variant discovery</li>
</ol>
<p>This step will combine the “gVCF” files and then call germline
variants in all samples. The paths of the local bundle files are also
required to be changed in the json template file. The final VCF file
of germline variants will be collected.</p>
<p>```</p>
</div>
<div id="gatk-alignment" class="section level2">
<h2><span class="header-section-number">13.4</span> GATK Alignment</h2>
<p>We wrapped the steps from raw fastqs to analysis-ready BAM file into <code>GAlign</code> pipeline. Here is the short summary of the pipeline.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(GAlign)
<span class="kw">short</span>(GAlign)</code></pre>
<pre><code>## inputs:
## - fastq1
## - fastq2
## - readGroup
## - sampleName
## - library
## - platunit
## - platform
## - center
## - tmpl1
## - wdl1
## - tmpl2
## - wdl2
## - cromwell
## outputs:
## - bamlog
## - outdir
## steps:
## - fqJson
## - fq2ubam
## - ubam2bamJson
## - align
## - mvOut</code></pre>
<p>For the <code>inputList</code>, we need to assign the fastqs files and read groups for each sample. The inputs can be multiple items separated by comma if there are more than one read groups for each sample. The input templates and WDL scripts can be assigned in the <code>paramList</code> and the reference and other GATK bundle files should be changed to your local version of files accordingly in the local json files. The path to the <code>cromwell</code> binary file is also required. Here is an example.</p>
<pre class="sourceCode r"><code class="sourceCode r">tmpl1 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/seq-format-conversion/paired-fastq-to-unmapped-bam.inputs.json&quot;</span>)
tmpl2 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-data-processing/processing-for-variant-discovery-gatk4.hg38.wgs.inputs.local.json&quot;</span>)
wdl1 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/seq-format-conversion/paired-fastq-to-unmapped-bam.wdl&quot;</span>)
wdl2 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-data-processing/processing-for-variant-discovery-gatk4.wdl&quot;</span>)

inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">fastq1=</span><span class="kw">list</span>(<span class="dt">normal=</span><span class="st">&quot;apps/DNASeq/data/normal_1.fq&quot;</span>,
                              <span class="dt">tumor=</span><span class="st">&quot;apps/DNASeq/data/tumor_1.fq&quot;</span>),
                  <span class="dt">fastq2=</span><span class="kw">list</span>(<span class="dt">normal=</span><span class="st">&quot;apps/DNASeq/data/normal_2.fq&quot;</span>,
                              <span class="dt">tumor=</span><span class="st">&quot;apps/DNASeq/data/tumor_2.fq&quot;</span>),
                  <span class="dt">readGroup=</span><span class="kw">list</span>(<span class="st">&quot;normal.1&quot;</span>, <span class="st">&quot;tumor.1&quot;</span>),
                  <span class="dt">sampleName=</span><span class="kw">list</span>(<span class="st">&quot;normal&quot;</span>, <span class="st">&quot;tumor&quot;</span>),
                  <span class="dt">library=</span><span class="kw">list</span>(<span class="st">&quot;normal&quot;</span>, <span class="st">&quot;tumor&quot;</span>),
                  <span class="dt">platunit=</span><span class="kw">list</span>(<span class="st">&quot;normal&quot;</span>, <span class="st">&quot;tumor&quot;</span>),
                  <span class="dt">platform=</span><span class="kw">list</span>(<span class="st">&quot;illumina&quot;</span>, <span class="st">&quot;illumina&quot;</span>),
                  <span class="dt">center=</span><span class="kw">list</span>(<span class="st">&quot;rpccc&quot;</span>, <span class="st">&quot;rpccc&quot;</span>))
paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">tmpl1=</span>tmpl1,
                  <span class="dt">wdl1=</span>wdl1,
                  <span class="dt">tmpl2=</span>tmpl2,
                  <span class="dt">wdl2=</span>wdl2,
                  <span class="dt">cromwell=</span><span class="st">&quot;/software/cromwell-36.jar&quot;</span>)
r1 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(GAlign, <span class="dt">outdir=</span><span class="st">&quot;apps/DNASeq/output/BAM&quot;</span>, inputList, paramList,
                  <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster=</span><span class="st">&quot;sge&quot;</span>,
                                  <span class="dt">template =</span> <span class="st">&quot;apps/data/sge_centos7.tmpl&quot;</span>,
                                  <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,
                                                   <span class="dt">queue =</span> <span class="st">&quot;centos7.q&quot;</span>)),
                  <span class="dt">stderr=</span><span class="st">&quot;&quot;</span>)</code></pre>
<p>The outputs were globbed from the <code>cromwell</code> execution folder.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list.files</span>(<span class="st">&quot;apps/DNASeq/output/BAM/normal&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</code></pre>
<pre><code>## [1] &quot;output/normal.hg38.bai&quot;                        
## [2] &quot;output/normal.hg38.bam&quot;                        
## [3] &quot;output/normal.hg38.bam.md5&quot;                    
## [4] &quot;output/normal.hg38.duplicate_metrics&quot;          
## [5] &quot;output/normal.hg38.recal_data.csv&quot;             
## [6] &quot;processing-for-variant-discovery-gatk4.wdl.log&quot;</code></pre>
</div>
<div id="haplotypecaller" class="section level2">
<h2><span class="header-section-number">13.5</span> HaplotypeCaller</h2>
<p>This step takes the BAM files as input and each BAM file will be assigned to different computing node. Also the json template file need to be modified to the correct GATK bundle paths at first.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(hapCall)
wdl3 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.wdl&quot;</span>)
tmpl3 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.hg38.inputs.local.json&quot;</span>)

bams &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">normal =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;output/BAM/normal/output/normal.hg38.bam&quot;</span>),
             <span class="dt">tumor =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;output/BAM/tumor/output/tumor.hg38.bam&quot;</span>))
inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">bam =</span> bams)
paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">intervals =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;output/interval.txt&quot;</span>),
                  <span class="dt">cromwell =</span> <span class="st">&quot;/software/cromwell-36.jar&quot;</span>,
                  <span class="dt">wdl =</span> wdl3,
                  <span class="dt">tmpl =</span> tmpl3)

r2 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(hapCall, <span class="dt">outdir=</span><span class="st">&quot;apps/DNASeq/output/GATK&quot;</span>, inputList, paramList,
                  <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster=</span><span class="st">&quot;sge&quot;</span>,
                                  <span class="dt">template =</span> <span class="st">&quot;apps/data/sge_centos7.tmpl&quot;</span>,
                                  <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,
                                                   <span class="dt">queue =</span> <span class="st">&quot;centos7.q&quot;</span>)),
                  <span class="dt">stderr=</span><span class="st">&quot;&quot;</span>)</code></pre>
<p>Here are the outputs:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list.files</span>(<span class="st">&quot;apps/DNASeq/output/GATK/normal&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</code></pre>
<pre><code>## [1] &quot;haplotypecaller-gvcf-gatk4.wdl.log&quot;
## [2] &quot;output/normal.hg38.g.vcf.gz&quot;       
## [3] &quot;output/normal.hg38.g.vcf.gz.tbi&quot;</code></pre>
</div>
<div id="joint-discovery" class="section level2">
<h2><span class="header-section-number">13.6</span> Joint Discovery</h2>
<p>The joint genotyping step will combine the <code>gvcf</code> files and then call variants in all samples, so only one computing node is required. Multiple values or files of the samples need to be pasted by comma for each input in the <code>inputList</code>. The paths of the local bundle files are also need to be added to the json template file.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(jdCall)
wdl4 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.wdl&quot;</span>)
tmpl4 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.hg38.wgs.inputs.json&quot;</span>)

inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sampleName =</span> <span class="kw">list</span>(<span class="dt">test=</span><span class="st">&quot;normal,tumor&quot;</span>),
                  <span class="dt">gvcf =</span> <span class="kw">list</span>(<span class="dt">test=</span><span class="st">&quot;apps/DNASeq/output/GATK/normal/output/normal.hg38.g.vcf.gz,apps/DNASeq/output/GATK/tumor/output/tumor.hg38.g.vcf.gz&quot;</span>))

paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">callsetName =</span> <span class="st">&quot;test&quot;</span>,
                  <span class="dt">intervals =</span> <span class="st">&quot;apps/DNASeq/output/interval.21.interval_list&quot;</span>,
                  <span class="dt">unpadded_intervals =</span> <span class="st">&quot;apps/DNASeq/output/interval.21.intervals&quot;</span>,
                  <span class="dt">tmpl =</span> tmpl4,
                  <span class="dt">cromwell =</span> <span class="st">&quot;/software/cromwell-36.jar&quot;</span>,
                  <span class="dt">wdl =</span> wdl4)

r3 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(jdCall, <span class="dt">outdir=</span><span class="st">&quot;apps/DNASeq/output/GATK&quot;</span>, inputList, paramList,
                  <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">1</span>, <span class="dt">cluster=</span><span class="st">&quot;sge&quot;</span>,
                                  <span class="dt">template =</span> <span class="st">&quot;apps/sge_centos7.tmpl&quot;</span>,
                                  <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,
                                                   <span class="dt">queue =</span> <span class="st">&quot;centos7.q&quot;</span>)),
                  <span class="dt">stderr=</span><span class="st">&quot;&quot;</span>)</code></pre>
<p>Here are the final outputs:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list.files</span>(<span class="st">&quot;apps/DNASeq/output/GATK/test&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</code></pre>
<pre><code>## [1] &quot;joint-discovery-gatk4-local.wdl.log&quot;        
## [2] &quot;output/out.intervals&quot;                       
## [3] &quot;output/test.variant_calling_detail_metrics&quot; 
## [4] &quot;output/test.variant_calling_summary_metrics&quot;
## [5] &quot;output/test.vcf.gz&quot;                         
## [6] &quot;output/test.vcf.gz.tbi&quot;</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="summarize-qc.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/05-application.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
