<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Neoantigen prediction | Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Neoantigen prediction | Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Neoantigen prediction | Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2019-10-11" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="rna-seq-alignment-and-quantification.html"/>
<link rel="next" href="other-sequencing-pipelines.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/d3-3.3.8/d3.min.js"></script>
<script src="libs/dagre-0.4.0/dagre-d3.min.js"></script>
<link href="libs/mermaid-0.3.0/dist/mermaid.css" rel="stylesheet" />
<script src="libs/mermaid-0.3.0/dist/mermaid.slim.min.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/chromatography-0.1/chromatography.js"></script>
<script src="libs/DiagrammeR-binding-1.0.1/DiagrammeR.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#common-workflow-language"><i class="fa fa-check"></i><b>1.1</b> Common Workflow Language</a></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#first-example"><i class="fa fa-check"></i><b>1.2</b> First example</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#test-run"><i class="fa fa-check"></i><b>1.3</b> Test run</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="components.html"><a href="components.html"><i class="fa fa-check"></i><b>2</b> Components</a><ul>
<li class="chapter" data-level="2.1" data-path="components.html"><a href="components.html#input-parameters"><i class="fa fa-check"></i><b>2.1</b> Input Parameters</a><ul>
<li class="chapter" data-level="2.1.1" data-path="components.html"><a href="components.html#essential-input-parameters"><i class="fa fa-check"></i><b>2.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="2.1.2" data-path="components.html"><a href="components.html#array-inputs"><i class="fa fa-check"></i><b>2.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="components.html"><a href="components.html#output-parameters"><i class="fa fa-check"></i><b>2.2</b> Output Parameters</a><ul>
<li class="chapter" data-level="2.2.1" data-path="components.html"><a href="components.html#capturing-output"><i class="fa fa-check"></i><b>2.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="2.2.2" data-path="components.html"><a href="components.html#array-outputs"><i class="fa fa-check"></i><b>2.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="2.2.3" data-path="components.html"><a href="components.html#standard-output"><i class="fa fa-check"></i><b>2.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>3</b> Writing Pipeline</a><ul>
<li class="chapter" data-level="3.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>3.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="3.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>3.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="run-approaches.html"><a href="run-approaches.html"><i class="fa fa-check"></i><b>4</b> Run approaches</a><ul>
<li class="chapter" data-level="4.1" data-path="run-approaches.html"><a href="run-approaches.html#running-tools-in-docker"><i class="fa fa-check"></i><b>4.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="4.2" data-path="run-approaches.html"><a href="run-approaches.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>4.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="4.3" data-path="run-approaches.html"><a href="run-approaches.html#web-application"><i class="fa fa-check"></i><b>4.3</b> Web Application</a><ul>
<li class="chapter" data-level="4.3.1" data-path="run-approaches.html"><a href="run-approaches.html#cwlparam-example"><i class="fa fa-check"></i><b>4.3.1</b> cwlParam example</a></li>
<li class="chapter" data-level="4.3.2" data-path="run-approaches.html"><a href="run-approaches.html#cwlparam-to-shiny-app"><i class="fa fa-check"></i><b>4.3.2</b> cwlParam to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="application.html"><a href="application.html"><i class="fa fa-check"></i><b>5</b> Application</a><ul>
<li class="chapter" data-level="5.1" data-path="application.html"><a href="application.html#rcwlpipelines-tools"><i class="fa fa-check"></i><b>5.1</b> RcwlPipelines tools</a><ul>
<li class="chapter" data-level="5.1.1" data-path="application.html"><a href="application.html#rcwl-scripts"><i class="fa fa-check"></i><b>5.1.1</b> Rcwl scripts</a></li>
<li class="chapter" data-level="5.1.2" data-path="application.html"><a href="application.html#build-a-pipeline"><i class="fa fa-check"></i><b>5.1.2</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="application.html"><a href="application.html#rcwlpipelines-summary"><i class="fa fa-check"></i><b>5.2</b> RcwlPipelines summary</a></li>
<li class="chapter" data-level="5.3" data-path="application.html"><a href="application.html#dnaseq-alignment-pipeline"><i class="fa fa-check"></i><b>5.3</b> DNASeq alignment pipeline</a><ul>
<li class="chapter" data-level="5.3.1" data-path="application.html"><a href="application.html#prepare-data"><i class="fa fa-check"></i><b>5.3.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.3.2" data-path="application.html"><a href="application.html#run-in-cluster"><i class="fa fa-check"></i><b>5.3.2</b> Run in cluster</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="application.html"><a href="application.html#rnaseq-pipeline"><i class="fa fa-check"></i><b>5.4</b> RNASeq pipeline</a><ul>
<li class="chapter" data-level="5.4.1" data-path="application.html"><a href="application.html#prepare-data-1"><i class="fa fa-check"></i><b>5.4.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.4.2" data-path="application.html"><a href="application.html#submit-pipeline-with-sge"><i class="fa fa-check"></i><b>5.4.2</b> Submit pipeline with SGE</a></li>
<li class="chapter" data-level="5.4.3" data-path="application.html"><a href="application.html#summarize-qc"><i class="fa fa-check"></i><b>5.4.3</b> Summarize QC</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="application.html"><a href="application.html#mc3-somatic-variant-calling-pipeline"><i class="fa fa-check"></i><b>5.5</b> MC3 somatic variant calling pipeline</a><ul>
<li class="chapter" data-level="5.5.1" data-path="application.html"><a href="application.html#prepare-data-2"><i class="fa fa-check"></i><b>5.5.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.5.2" data-path="application.html"><a href="application.html#run-mc3-pipeline"><i class="fa fa-check"></i><b>5.5.2</b> Run MC3 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="application.html"><a href="application.html#gatk4-germline-variant-calling-pipeline"><i class="fa fa-check"></i><b>5.6</b> GATK4 germline variant calling pipeline</a><ul>
<li class="chapter" data-level="5.6.1" data-path="application.html"><a href="application.html#gatk-alignment"><i class="fa fa-check"></i><b>5.6.1</b> GATK Alignment</a></li>
<li class="chapter" data-level="5.6.2" data-path="application.html"><a href="application.html#haplotypecaller"><i class="fa fa-check"></i><b>5.6.2</b> HaplotypeCaller</a></li>
<li class="chapter" data-level="5.6.3" data-path="application.html"><a href="application.html#joint-discovery"><i class="fa fa-check"></i><b>5.6.3</b> Joint Discovery</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html"><i class="fa fa-check"></i><b>6</b> DNA-Seq Variant calling</a><ul>
<li class="chapter" data-level="6.1" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#dna-alignment"><i class="fa fa-check"></i><b>6.1</b> DNA alignment</a></li>
<li class="chapter" data-level="6.2" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#germline-variant-calling"><i class="fa fa-check"></i><b>6.2</b> Germline variant calling</a></li>
<li class="chapter" data-level="6.3" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#somatic-mutation-calling"><i class="fa fa-check"></i><b>6.3</b> Somatic mutation calling</a><ul>
<li class="chapter" data-level="6.3.1" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#mutect2"><i class="fa fa-check"></i><b>6.3.1</b> MuTect2</a></li>
<li class="chapter" data-level="6.3.2" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#muse"><i class="fa fa-check"></i><b>6.3.2</b> MuSE</a></li>
<li class="chapter" data-level="6.3.3" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#strelka2"><i class="fa fa-check"></i><b>6.3.3</b> Strelka2</a></li>
<li class="chapter" data-level="6.3.4" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#somaticsniper"><i class="fa fa-check"></i><b>6.3.4</b> SomaticSniper</a></li>
<li class="chapter" data-level="6.3.5" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#vardict"><i class="fa fa-check"></i><b>6.3.5</b> VarDict</a></li>
<li class="chapter" data-level="6.3.6" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#lofreq"><i class="fa fa-check"></i><b>6.3.6</b> LoFreq</a></li>
<li class="chapter" data-level="6.3.7" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#varscan2"><i class="fa fa-check"></i><b>6.3.7</b> VarScan2</a></li>
<li class="chapter" data-level="6.3.8" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#lancet"><i class="fa fa-check"></i><b>6.3.8</b> Lancet</a></li>
<li class="chapter" data-level="6.3.9" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#neusomatic"><i class="fa fa-check"></i><b>6.3.9</b> Neusomatic</a></li>
<li class="chapter" data-level="6.3.10" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#combined-somatic-caller-pipeline"><i class="fa fa-check"></i><b>6.3.10</b> Combined somatic caller pipeline</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#structural-variation"><i class="fa fa-check"></i><b>6.4</b> Structural variation</a><ul>
<li class="chapter" data-level="6.4.1" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#structural-variation-engine"><i class="fa fa-check"></i><b>6.4.1</b> Structural variation Engine</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#copy-number-variation"><i class="fa fa-check"></i><b>6.5</b> Copy number variation</a><ul>
<li class="chapter" data-level="6.5.1" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#cnvkit"><i class="fa fa-check"></i><b>6.5.1</b> cnvkit</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html"><i class="fa fa-check"></i><b>7</b> RNA-Seq alignment and quantification</a><ul>
<li class="chapter" data-level="7.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#bulk-mrna"><i class="fa fa-check"></i><b>7.1</b> bulk mRNA</a><ul>
<li class="chapter" data-level="7.1.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#prepare-data-3"><i class="fa fa-check"></i><b>7.1.1</b> Prepare data</a></li>
<li class="chapter" data-level="7.1.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#submit-pipeline-with-sge-1"><i class="fa fa-check"></i><b>7.1.2</b> Submit pipeline with SGE</a></li>
<li class="chapter" data-level="7.1.3" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#qc-summary"><i class="fa fa-check"></i><b>7.1.3</b> QC Summary</a></li>
<li class="chapter" data-level="7.1.4" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#abundances-summary"><i class="fa fa-check"></i><b>7.1.4</b> Abundances summary</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#transcriptome-quantification"><i class="fa fa-check"></i><b>7.2</b> transcriptome quantification</a><ul>
<li class="chapter" data-level="7.2.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#kallisto"><i class="fa fa-check"></i><b>7.2.1</b> Kallisto</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#single-cell-rnaseq"><i class="fa fa-check"></i><b>7.3</b> single cell RNAseq</a></li>
<li class="chapter" data-level="7.4" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#mirna"><i class="fa fa-check"></i><b>7.4</b> miRNA</a><ul>
<li class="chapter" data-level="7.4.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#prepare-reference"><i class="fa fa-check"></i><b>7.4.1</b> Prepare reference</a></li>
<li class="chapter" data-level="7.4.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>7.4.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html"><i class="fa fa-check"></i><b>8</b> Neoantigen prediction</a><ul>
<li class="chapter" data-level="8.1" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#input-file-preparation"><i class="fa fa-check"></i><b>8.1</b> Input file preparation</a><ul>
<li class="chapter" data-level="8.1.1" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#annotating-vcf"><i class="fa fa-check"></i><b>8.1.1</b> Annotating VCF</a></li>
<li class="chapter" data-level="8.1.2" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#adding-coverage"><i class="fa fa-check"></i><b>8.1.2</b> Adding coverage</a></li>
<li class="chapter" data-level="8.1.3" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#adding-expression"><i class="fa fa-check"></i><b>8.1.3</b> Adding expression</a></li>
<li class="chapter" data-level="8.1.4" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#creating-a-phased-vcf"><i class="fa fa-check"></i><b>8.1.4</b> Creating a phased VCF</a></li>
<li class="chapter" data-level="8.1.5" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#combined-vcf-preparing-pipeline"><i class="fa fa-check"></i><b>8.1.5</b> Combined VCF preparing pipeline</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#hla-typing"><i class="fa fa-check"></i><b>8.2</b> HLA typing</a><ul>
<li class="chapter" data-level="8.2.1" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#polysolver"><i class="fa fa-check"></i><b>8.2.1</b> polysolver</a></li>
<li class="chapter" data-level="8.2.2" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#optitype"><i class="fa fa-check"></i><b>8.2.2</b> Optitype</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="neoantigen-prediction.html"><a href="neoantigen-prediction.html#neoantigen-prediction-by-pvacseq"><i class="fa fa-check"></i><b>8.3</b> Neoantigen prediction by pVACseq</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-sequencing-pipelines.html"><a href="other-sequencing-pipelines.html"><i class="fa fa-check"></i><b>9</b> Other Sequencing pipelines</a><ul>
<li class="chapter" data-level="9.1" data-path="other-sequencing-pipelines.html"><a href="other-sequencing-pipelines.html#chip-seq"><i class="fa fa-check"></i><b>9.1</b> ChiP-Seq</a></li>
<li class="chapter" data-level="9.2" data-path="other-sequencing-pipelines.html"><a href="other-sequencing-pipelines.html#atac-seq"><i class="fa fa-check"></i><b>9.2</b> ATAC-Seq</a></li>
<li class="chapter" data-level="9.3" data-path="other-sequencing-pipelines.html"><a href="other-sequencing-pipelines.html#microbiome-sequencing"><i class="fa fa-check"></i><b>9.3</b> Microbiome Sequencing</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="neoantigen-prediction" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Neoantigen prediction</h1>
<p>The “pVACseq” toolkit is a comprehensive cancer immunotherapy pipeline for identification of personalized variant antigens by integrating with DNA mutations and mRNA expression data to filter and rank candicate neoepitopes. The “POLYSOLVER” or “Optitype” is a tool for HLA typing. It can be used to limit the list of alleles.</p>
<p>The full pipeline is shown as follows.
<div id="htmlwidget-11cc50c170ecff74102c" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-11cc50c170ecff74102c">{"x":{"diagram":"\ngraph LR\n  A(tumor DNA reads)-->D{Alignment}\n  B(normal DNA reads)-->D{Alignment}\n  C(tumor RNA reads)-->E{Quantification}\n  D{Alignment}-->F((tumor BAM))\n  D{Alignment}-->G((normal BAM))\n  F((tumor BAM))-->H{Somatic variant calling}\n  G((normal BAM))-->H{Somatic variant calling}\n  G((normal BAM))-->I{Germline variant calling}\n  H{Somatic variant calling}-->J((Somatic VCF))\n  I{Germline variant calling}-->K((Germline VCF))\n  J((Somatic VCF))-->L{Annotating and Phasing}\n  K((Germline VCF))-->L{Annotating and Phasing}\n  E{Quantification}-->L{Annotating and Phasing}\n  L{Annotation and Phasing}-->M((annotated VCF))\n  L{Annotation and Phasing}-->N((phased VCF))\n  F((tumor BAM))-->O{HLA typing}\n  M((annotated VCF))-->P{Neoantigen prediction}\n  N((phased VCF))-->P{Neoantigen prediction}\n  O{HLA typing}-->P{Neoantigen prediction}\n  P{Neoantigen prediction}-->Q((Ranked neoantigens))\n"},"evals":[],"jsHooks":[]}</script></p>
<p>Three major steps are built to run the full pipeline, including DNA variant calling, variant annotation and neoantigen prediction.</p>
<ul>
<li>DNA variant calling: somatic and germline variant calling</li>
<li>Variant annotation: Adding DNA and RNA allele frequencies, transcript and gene level expressions.</li>
<li>Neoantigen prediction: HLA typing, neoantigen filtering and ranking</li>
</ul>
<p>First of all, load the required packages.</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="neoantigen-prediction.html#cb265-1"></a><span class="kw">library</span>(RcwlPipelines)</span>
<span id="cb265-2"><a href="neoantigen-prediction.html#cb265-2"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb265-3"><a href="neoantigen-prediction.html#cb265-3"></a><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(VariantAnnotation))</span>
<span id="cb265-4"><a href="neoantigen-prediction.html#cb265-4"></a><span class="kw">library</span>(fs)</span>
<span id="cb265-5"><a href="neoantigen-prediction.html#cb265-5"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb265-6"><a href="neoantigen-prediction.html#cb265-6"></a><span class="kw">library</span>(conflicted)</span>
<span id="cb265-7"><a href="neoantigen-prediction.html#cb265-7"></a><span class="co">## always use select from dplyr</span></span>
<span id="cb265-8"><a href="neoantigen-prediction.html#cb265-8"></a><span class="kw">conflict_prefer</span>(<span class="st">&quot;select&quot;</span>, <span class="st">&quot;dplyr&quot;</span>, <span class="dt">quiet =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div id="input-file-preparation" class="section level2">
<h2><span class="header-section-number">8.1</span> Input file preparation</h2>
<div id="annotating-vcf" class="section level3">
<h3><span class="header-section-number">8.1.1</span> Annotating VCF</h3>
<p>First, we need to annotate somatic mutations using the tool “vep” with “Downstream” and “Wildtype” plugins. A cache of annotation sources is recommended to use “VEP” in the fastest and most efficient way. <a href="http://useast.ensembl.org/info/docs/tools/vep/script/vep_cache.html" class="uri">http://useast.ensembl.org/info/docs/tools/vep/script/vep_cache.html</a></p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="neoantigen-prediction.html#cb266-1"></a>ref &lt;-<span class="st"> &quot;apps/data/hs37d5.fa&quot;</span></span>
<span id="cb266-2"><a href="neoantigen-prediction.html#cb266-2"></a>vcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/combined&quot;</span>, <span class="st">&quot;.neusomatic.vcf&quot;</span>,</span>
<span id="cb266-3"><a href="neoantigen-prediction.html#cb266-3"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb266-4"><a href="neoantigen-prediction.html#cb266-4"></a><span class="kw">names</span>(vcfs) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb266-5"><a href="neoantigen-prediction.html#cb266-5"></a>ovcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">paste0</span>(<span class="kw">names</span>(vcfs), <span class="st">&quot;_vep.vcf&quot;</span>))</span>
<span id="cb266-6"><a href="neoantigen-prediction.html#cb266-6"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ivcf =</span> vcfs,</span>
<span id="cb266-7"><a href="neoantigen-prediction.html#cb266-7"></a>                  <span class="dt">ovcf =</span> ovcfs)</span>
<span id="cb266-8"><a href="neoantigen-prediction.html#cb266-8"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ref =</span> ref,</span>
<span id="cb266-9"><a href="neoantigen-prediction.html#cb266-9"></a>                  <span class="dt">cacheDir =</span> <span class="st">&quot;~/.vep&quot;</span>)</span>
<span id="cb266-10"><a href="neoantigen-prediction.html#cb266-10"></a></span>
<span id="cb266-11"><a href="neoantigen-prediction.html#cb266-11"></a>res1 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(vep, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb266-12"><a href="neoantigen-prediction.html#cb266-12"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb266-13"><a href="neoantigen-prediction.html#cb266-13"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb266-14"><a href="neoantigen-prediction.html#cb266-14"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb266-15"><a href="neoantigen-prediction.html#cb266-15"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb266-16"><a href="neoantigen-prediction.html#cb266-16"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb266-17"><a href="neoantigen-prediction.html#cb266-17"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;vep&quot;</span>),</span>
<span id="cb266-18"><a href="neoantigen-prediction.html#cb266-18"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb266-19"><a href="neoantigen-prediction.html#cb266-19"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Here are the annotated VCFs.</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="neoantigen-prediction.html#cb267-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>, <span class="dt">glob =</span> <span class="st">&quot;*vep.vcf&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 2 x 2
##   path                                              size
##   &lt;fs::path&gt;                                 &lt;fs::bytes&gt;
## 1 apps/variants/neoantigen/pt19/pt19_vep.vcf        642K
## 2 apps/variants/neoantigen/pt20/pt20_vep.vcf        217K</code></pre>
</div>
<div id="adding-coverage" class="section level3">
<h3><span class="header-section-number">8.1.2</span> Adding coverage</h3>
<p>This step use <code>bam-readcount</code> to to retrieve the number of reads in given VCF positions. Then <code>vcf-readcount-annotator</code> is used to add the read counts for snvs and indels to the input VCF file.</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="neoantigen-prediction.html#cb269-1"></a><span class="kw">plotCWL</span>(vcfCoverage)</span></code></pre></div>
<div id="htmlwidget-355da6228f271c5ead0c" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-355da6228f271c5ead0c">{"x":{"diagram":"graph TB\nA(vcf)-->G{decompose}\nG{decompose}-->|oVcf|H{readcount}\nB(sample)-->H{readcount}\nE(ref)-->H{readcount}\nC(bam)-->H{readcount}\nG{decompose}-->|oVcf|I{readcount_annotator_snv}\nH{readcount}-->|snv|I{readcount_annotator_snv}\nD(ntype)-->I{readcount_annotator_snv}\nB(sample)-->I{readcount_annotator_snv}\nI{readcount_annotator_snv}-->|oVcf|J{readcount_annotator_indel}\nH{readcount}-->|indel|J{readcount_annotator_indel}\nD(ntype)-->J{readcount_annotator_indel}\nB(sample)-->J{readcount_annotator_indel}\nJ{readcount_annotator_indel}-->F((outvcf))"},"evals":[],"jsHooks":[]}</script>
<p>Let’s prepare the inputs for DNA samples and run the <code>vcfCoverage</code> pipeline.</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="neoantigen-prediction.html#cb270-1"></a>vcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;*vep.vcf&quot;</span>,</span>
<span id="cb270-2"><a href="neoantigen-prediction.html#cb270-2"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb270-3"><a href="neoantigen-prediction.html#cb270-3"></a>bams &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/BAM&quot;</span>, <span class="st">&quot;tumor.*.bam$&quot;</span>,</span>
<span id="cb270-4"><a href="neoantigen-prediction.html#cb270-4"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb270-5"><a href="neoantigen-prediction.html#cb270-5"></a>samples  &lt;-<span class="st">  </span><span class="kw">as.list</span>(<span class="kw">c</span>(<span class="st">&quot;SAMPLE&quot;</span>, <span class="st">&quot;SAMPLE&quot;</span>))</span>
<span id="cb270-6"><a href="neoantigen-prediction.html#cb270-6"></a><span class="kw">names</span>(vcfs) &lt;-<span class="st"> </span><span class="kw">names</span>(bams) &lt;-<span class="st"> </span><span class="kw">names</span>(samples) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb270-7"><a href="neoantigen-prediction.html#cb270-7"></a></span>
<span id="cb270-8"><a href="neoantigen-prediction.html#cb270-8"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">vcf =</span> vcfs,</span>
<span id="cb270-9"><a href="neoantigen-prediction.html#cb270-9"></a>                  <span class="dt">sample =</span> samples,</span>
<span id="cb270-10"><a href="neoantigen-prediction.html#cb270-10"></a>                  <span class="dt">bam =</span> bams)</span>
<span id="cb270-11"><a href="neoantigen-prediction.html#cb270-11"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ref =</span> ref)</span>
<span id="cb270-12"><a href="neoantigen-prediction.html#cb270-12"></a></span>
<span id="cb270-13"><a href="neoantigen-prediction.html#cb270-13"></a>res2 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(vcfCoverage, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb270-14"><a href="neoantigen-prediction.html#cb270-14"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb270-15"><a href="neoantigen-prediction.html#cb270-15"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb270-16"><a href="neoantigen-prediction.html#cb270-16"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb270-17"><a href="neoantigen-prediction.html#cb270-17"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb270-18"><a href="neoantigen-prediction.html#cb270-18"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb270-19"><a href="neoantigen-prediction.html#cb270-19"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;coverage&quot;</span>),</span>
<span id="cb270-20"><a href="neoantigen-prediction.html#cb270-20"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb270-21"><a href="neoantigen-prediction.html#cb270-21"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The Allelic depths, “AD”, for the ref and alt alleles were added to the VCF file.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="neoantigen-prediction.html#cb271-1"></a>vcf1 &lt;-<span class="st"> </span><span class="kw">readVcf</span>(<span class="st">&quot;apps/variants/neoantigen/pt19/pt19_vep_dc_snv_indel.vcf&quot;</span>)</span>
<span id="cb271-2"><a href="neoantigen-prediction.html#cb271-2"></a><span class="kw">geno</span>(vcf1)</span></code></pre></div>
<pre><code>## List of length 6
## names(6): GT RO AO DP AD AF</code></pre>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="neoantigen-prediction.html#cb273-1"></a><span class="kw">do.call</span>(rbind, <span class="kw">head</span>(<span class="kw">geno</span>(vcf1)<span class="op">$</span>AD))</span></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]   34    8
## [2,]   37   11
## [3,]   20   19
## [4,]   32   16
## [5,]    6    4
## [6,]   18   18</code></pre>
<p>We also could run the pipeline for RNA samples with “ntype=‘RNA’”. Here we don’t have their RNA BAM files, so the DNA BAM files are used for demonstration.</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="neoantigen-prediction.html#cb275-1"></a>vcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;*vep_dc_snv_indel.vcf&quot;</span>,</span>
<span id="cb275-2"><a href="neoantigen-prediction.html#cb275-2"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb275-3"><a href="neoantigen-prediction.html#cb275-3"></a>bams &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/BAM&quot;</span>, <span class="st">&quot;tumor.*.bam$&quot;</span>,</span>
<span id="cb275-4"><a href="neoantigen-prediction.html#cb275-4"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb275-5"><a href="neoantigen-prediction.html#cb275-5"></a>samples  &lt;-<span class="st">  </span><span class="kw">as.list</span>(<span class="kw">c</span>(<span class="st">&quot;SAMPLE&quot;</span>, <span class="st">&quot;SAMPLE&quot;</span>))</span>
<span id="cb275-6"><a href="neoantigen-prediction.html#cb275-6"></a><span class="kw">names</span>(vcfs) &lt;-<span class="st"> </span><span class="kw">names</span>(bams) &lt;-<span class="st"> </span><span class="kw">names</span>(samples) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb275-7"><a href="neoantigen-prediction.html#cb275-7"></a></span>
<span id="cb275-8"><a href="neoantigen-prediction.html#cb275-8"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">vcf =</span> vcfs,</span>
<span id="cb275-9"><a href="neoantigen-prediction.html#cb275-9"></a>                  <span class="dt">sample =</span> samples,</span>
<span id="cb275-10"><a href="neoantigen-prediction.html#cb275-10"></a>                  <span class="dt">bam =</span> bams)</span>
<span id="cb275-11"><a href="neoantigen-prediction.html#cb275-11"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ref =</span> ref,</span>
<span id="cb275-12"><a href="neoantigen-prediction.html#cb275-12"></a>                  <span class="dt">ntype =</span> <span class="st">&quot;RNA&quot;</span>)</span>
<span id="cb275-13"><a href="neoantigen-prediction.html#cb275-13"></a></span>
<span id="cb275-14"><a href="neoantigen-prediction.html#cb275-14"></a>res2a &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(vcfCoverage, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb275-15"><a href="neoantigen-prediction.html#cb275-15"></a>                     <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb275-16"><a href="neoantigen-prediction.html#cb275-16"></a>                                               <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb275-17"><a href="neoantigen-prediction.html#cb275-17"></a>                                               <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb275-18"><a href="neoantigen-prediction.html#cb275-18"></a>                                               <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb275-19"><a href="neoantigen-prediction.html#cb275-19"></a>                                                                <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb275-20"><a href="neoantigen-prediction.html#cb275-20"></a>                                                                <span class="dt">jobname =</span> <span class="st">&quot;coverage&quot;</span>),</span>
<span id="cb275-21"><a href="neoantigen-prediction.html#cb275-21"></a>                                               <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb275-22"><a href="neoantigen-prediction.html#cb275-22"></a>                     <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The Allelic depths, “RAD”, for the ref and alt alleles were added to the VCF file.</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="neoantigen-prediction.html#cb276-1"></a>vcf1a &lt;-<span class="st"> </span><span class="kw">readVcf</span>(<span class="st">&quot;apps/variants/neoantigen/pt19/pt19_vep_dc_snv_indel_dc_snv_indel.vcf&quot;</span>)</span>
<span id="cb276-2"><a href="neoantigen-prediction.html#cb276-2"></a><span class="kw">geno</span>(vcf1a)</span></code></pre></div>
<pre><code>## List of length 9
## names(9): GT RO AO DP AD AF RDP RAD RAF</code></pre>
</div>
<div id="adding-expression" class="section level3">
<h3><span class="header-section-number">8.1.3</span> Adding expression</h3>
<p>For demostration purpose, we use the same DNA fastq files as RNASeq inputs for transcriptome quantification with <code>kallisto_quant</code>.</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="neoantigen-prediction.html#cb278-1"></a>fqs &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&quot;apps/variants/fastq&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb278-2"><a href="neoantigen-prediction.html#cb278-2"></a>fqs &lt;-<span class="st"> </span>fqs[<span class="kw">grep</span>(<span class="st">&quot;tumor&quot;</span>, fqs)]</span>
<span id="cb278-3"><a href="neoantigen-prediction.html#cb278-3"></a>fqL &lt;-<span class="st"> </span><span class="kw">tapply</span>(fqs, <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>), <span class="dt">each =</span> <span class="dv">2</span>), as.list)</span>
<span id="cb278-4"><a href="neoantigen-prediction.html#cb278-4"></a></span>
<span id="cb278-5"><a href="neoantigen-prediction.html#cb278-5"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">fastq =</span> fqL)</span>
<span id="cb278-6"><a href="neoantigen-prediction.html#cb278-6"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">index =</span> <span class="st">&quot;apps/data/gencode.v25.transcripts_kallisto&quot;</span>,</span>
<span id="cb278-7"><a href="neoantigen-prediction.html#cb278-7"></a>                  <span class="dt">threads =</span> <span class="dv">16</span>)</span>
<span id="cb278-8"><a href="neoantigen-prediction.html#cb278-8"></a></span>
<span id="cb278-9"><a href="neoantigen-prediction.html#cb278-9"></a>res3 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(kallisto_quant, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb278-10"><a href="neoantigen-prediction.html#cb278-10"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="kw">length</span>(samples),</span>
<span id="cb278-11"><a href="neoantigen-prediction.html#cb278-11"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb278-12"><a href="neoantigen-prediction.html#cb278-12"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb278-13"><a href="neoantigen-prediction.html#cb278-13"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb278-14"><a href="neoantigen-prediction.html#cb278-14"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb278-15"><a href="neoantigen-prediction.html#cb278-15"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;kallisto&quot;</span>),</span>
<span id="cb278-16"><a href="neoantigen-prediction.html#cb278-16"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb278-17"><a href="neoantigen-prediction.html#cb278-17"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The tool <code>vcf_expression_annotator</code> can be used to parse gene or transcript expression results and add to a given VCF.</p>
<p>The abundance tables from <code>kallisto quant</code> have full transcript IDs with gene IDs annotation in the “target_id” column. We need to clean the “targe_id” column by removing the annotation parts other than “transcript_id”.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="neoantigen-prediction.html#cb279-1"></a>cleanExp &lt;-<span class="st"> </span><span class="cf">function</span>(afile) {</span>
<span id="cb279-2"><a href="neoantigen-prediction.html#cb279-2"></a>    exp1 &lt;-<span class="st"> </span><span class="kw">read.table</span>(afile, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb279-3"><a href="neoantigen-prediction.html#cb279-3"></a>    exp1[,<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|ENSG.*&quot;</span>, <span class="st">&quot;&quot;</span>, exp1[,<span class="dv">1</span>])</span>
<span id="cb279-4"><a href="neoantigen-prediction.html#cb279-4"></a>    <span class="kw">write.table</span>(exp1, <span class="dt">file =</span> <span class="st">&quot;abundance_clean.tsv&quot;</span>,</span>
<span id="cb279-5"><a href="neoantigen-prediction.html#cb279-5"></a>                <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span>
<span id="cb279-6"><a href="neoantigen-prediction.html#cb279-6"></a>}</span>
<span id="cb279-7"><a href="neoantigen-prediction.html#cb279-7"></a>p1 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;afile&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>,</span>
<span id="cb279-8"><a href="neoantigen-prediction.html#cb279-8"></a>                 <span class="dt">prefix =</span> <span class="st">&quot;afile=&quot;</span>, <span class="dt">separate =</span> <span class="ot">FALSE</span>)</span>
<span id="cb279-9"><a href="neoantigen-prediction.html#cb279-9"></a>o1 &lt;-<span class="st"> </span><span class="kw">OutputParam</span>(<span class="dt">id =</span> <span class="st">&quot;aout&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>, <span class="dt">glob =</span> <span class="st">&quot;abundance_clean.tsv&quot;</span>)</span>
<span id="cb279-10"><a href="neoantigen-prediction.html#cb279-10"></a>CleanExp &lt;-<span class="st"> </span><span class="kw">cwlParam</span>(<span class="dt">baseCommand =</span> cleanExp,</span>
<span id="cb279-11"><a href="neoantigen-prediction.html#cb279-11"></a>                     <span class="dt">inputs =</span> <span class="kw">InputParamList</span>(p1),</span>
<span id="cb279-12"><a href="neoantigen-prediction.html#cb279-12"></a>                     <span class="dt">outputs =</span> <span class="kw">OutputParamList</span>(o1))</span>
<span id="cb279-13"><a href="neoantigen-prediction.html#cb279-13"></a></span>
<span id="cb279-14"><a href="neoantigen-prediction.html#cb279-14"></a>exps &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;abundance.tsv&quot;</span>,</span>
<span id="cb279-15"><a href="neoantigen-prediction.html#cb279-15"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb279-16"><a href="neoantigen-prediction.html#cb279-16"></a><span class="kw">names</span>(exps) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb279-17"><a href="neoantigen-prediction.html#cb279-17"></a></span>
<span id="cb279-18"><a href="neoantigen-prediction.html#cb279-18"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">afile =</span> exps)</span>
<span id="cb279-19"><a href="neoantigen-prediction.html#cb279-19"></a>res4 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(CleanExp, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList,</span>
<span id="cb279-20"><a href="neoantigen-prediction.html#cb279-20"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb279-21"><a href="neoantigen-prediction.html#cb279-21"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb279-22"><a href="neoantigen-prediction.html#cb279-22"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb279-23"><a href="neoantigen-prediction.html#cb279-23"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb279-24"><a href="neoantigen-prediction.html#cb279-24"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb279-25"><a href="neoantigen-prediction.html#cb279-25"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;vep&quot;</span>),</span>
<span id="cb279-26"><a href="neoantigen-prediction.html#cb279-26"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb279-27"><a href="neoantigen-prediction.html#cb279-27"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Then the cleaned abundance file can be used to add expression values to the VEP annotated VCF file using the tool <code>vcf_expression_annotator</code>.</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="neoantigen-prediction.html#cb280-1"></a>vcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;*indel_dc_snv_indel.vcf&quot;</span>,</span>
<span id="cb280-2"><a href="neoantigen-prediction.html#cb280-2"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb280-3"><a href="neoantigen-prediction.html#cb280-3"></a>exps &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;abundance_clean.tsv&quot;</span>,</span>
<span id="cb280-4"><a href="neoantigen-prediction.html#cb280-4"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb280-5"><a href="neoantigen-prediction.html#cb280-5"></a>ovcf &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;pt19_Ann.vcf&quot;</span>, <span class="st">&quot;pt20_Ann.vcf&quot;</span>)</span>
<span id="cb280-6"><a href="neoantigen-prediction.html#cb280-6"></a><span class="kw">names</span>(vcfs) &lt;-<span class="st"> </span><span class="kw">names</span>(exps) &lt;-<span class="st"> </span><span class="kw">names</span>(ovcf) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb280-7"><a href="neoantigen-prediction.html#cb280-7"></a></span>
<span id="cb280-8"><a href="neoantigen-prediction.html#cb280-8"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ivcf =</span> vcfs,</span>
<span id="cb280-9"><a href="neoantigen-prediction.html#cb280-9"></a>                  <span class="dt">expression =</span> exps,</span>
<span id="cb280-10"><a href="neoantigen-prediction.html#cb280-10"></a>                  <span class="dt">ovcf =</span> ovcf)</span>
<span id="cb280-11"><a href="neoantigen-prediction.html#cb280-11"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">etype =</span> <span class="st">&quot;kallisto&quot;</span>,</span>
<span id="cb280-12"><a href="neoantigen-prediction.html#cb280-12"></a>                  <span class="dt">gtype =</span> <span class="st">&quot;transcript&quot;</span>)</span>
<span id="cb280-13"><a href="neoantigen-prediction.html#cb280-13"></a></span>
<span id="cb280-14"><a href="neoantigen-prediction.html#cb280-14"></a>res5 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(vcf_expression_annotator, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb280-15"><a href="neoantigen-prediction.html#cb280-15"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb280-16"><a href="neoantigen-prediction.html#cb280-16"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb280-17"><a href="neoantigen-prediction.html#cb280-17"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb280-18"><a href="neoantigen-prediction.html#cb280-18"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb280-19"><a href="neoantigen-prediction.html#cb280-19"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb280-20"><a href="neoantigen-prediction.html#cb280-20"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;vep&quot;</span>),</span>
<span id="cb280-21"><a href="neoantigen-prediction.html#cb280-21"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb280-22"><a href="neoantigen-prediction.html#cb280-22"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The transcript expression values, “TX”, were added to the VCF file.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="neoantigen-prediction.html#cb281-1"></a>vcf2 &lt;-<span class="st"> </span><span class="kw">readVcf</span>(<span class="st">&quot;apps/variants/neoantigen/pt19/pt19_Ann.vcf&quot;</span>)</span>
<span id="cb281-2"><a href="neoantigen-prediction.html#cb281-2"></a><span class="kw">geno</span>(vcf2)</span></code></pre></div>
<pre><code>## List of length 7
## names(7): GT RO AO DP AD AF TX</code></pre>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="neoantigen-prediction.html#cb283-1"></a><span class="kw">geno</span>(vcf2)<span class="op">$</span>TX[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## [1] &quot;ENST00000264554|31.1948&quot; &quot;ENST00000590113|203.734&quot;
## [3] &quot;ENST00000590170|41.8456&quot; &quot;ENST00000590222|101.681&quot;</code></pre>
<p>The gene level expression</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="neoantigen-prediction.html#cb285-1"></a><span class="kw">library</span>(tximport)</span>
<span id="cb285-2"><a href="neoantigen-prediction.html#cb285-2"></a></span>
<span id="cb285-3"><a href="neoantigen-prediction.html#cb285-3"></a>t2gene &lt;-<span class="st"> </span><span class="cf">function</span>(kexp){</span>
<span id="cb285-4"><a href="neoantigen-prediction.html#cb285-4"></a>    e1 &lt;-<span class="st"> </span><span class="kw">read.table</span>(kexp, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">check.names =</span> <span class="ot">FALSE</span>,</span>
<span id="cb285-5"><a href="neoantigen-prediction.html#cb285-5"></a>                     <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span>
<span id="cb285-6"><a href="neoantigen-prediction.html#cb285-6"></a>    ids &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, base<span class="op">::</span><span class="kw">strsplit</span>(e1<span class="op">$</span>target_id, <span class="dt">split =</span> <span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))</span>
<span id="cb285-7"><a href="neoantigen-prediction.html#cb285-7"></a>    tx2gene  &lt;-<span class="st"> </span><span class="kw">data.frame</span>(ids[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</span>
<span id="cb285-8"><a href="neoantigen-prediction.html#cb285-8"></a>    gexp &lt;-<span class="st"> </span>tximport<span class="op">::</span><span class="kw">tximport</span>(kexp, <span class="dt">type =</span> <span class="st">&quot;kallisto&quot;</span>, <span class="dt">tx2gene =</span> tx2gene, <span class="dt">ignoreAfterBar=</span><span class="ot">TRUE</span>)</span>
<span id="cb285-9"><a href="neoantigen-prediction.html#cb285-9"></a>    gExp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene =</span> <span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">..*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">rownames</span>(gexp<span class="op">$</span>abundance)),</span>
<span id="cb285-10"><a href="neoantigen-prediction.html#cb285-10"></a>                       <span class="dt">abundance =</span> gexp<span class="op">$</span>abundance)</span>
<span id="cb285-11"><a href="neoantigen-prediction.html#cb285-11"></a>    <span class="kw">write.table</span>(gExp, <span class="dt">file =</span> <span class="st">&quot;abundance_gene.tsv&quot;</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>,</span>
<span id="cb285-12"><a href="neoantigen-prediction.html#cb285-12"></a>                <span class="dt">col.names =</span> <span class="ot">TRUE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span>
<span id="cb285-13"><a href="neoantigen-prediction.html#cb285-13"></a>}</span>
<span id="cb285-14"><a href="neoantigen-prediction.html#cb285-14"></a>p1 &lt;-<span class="st"> </span><span class="kw">InputParam</span>(<span class="dt">id =</span> <span class="st">&quot;kexp&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>,</span>
<span id="cb285-15"><a href="neoantigen-prediction.html#cb285-15"></a>                 <span class="dt">prefix =</span> <span class="st">&quot;kexp=&quot;</span>, <span class="dt">separate =</span> <span class="ot">FALSE</span>)</span>
<span id="cb285-16"><a href="neoantigen-prediction.html#cb285-16"></a>o1 &lt;-<span class="st"> </span><span class="kw">OutputParam</span>(<span class="dt">id =</span> <span class="st">&quot;gout&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;File&quot;</span>, <span class="dt">glob =</span> <span class="st">&quot;abundance_gene.tsv&quot;</span>)</span>
<span id="cb285-17"><a href="neoantigen-prediction.html#cb285-17"></a>T2Gene &lt;-<span class="st"> </span><span class="kw">cwlParam</span>(<span class="dt">baseCommand =</span> t2gene,</span>
<span id="cb285-18"><a href="neoantigen-prediction.html#cb285-18"></a>                   <span class="dt">inputs =</span> <span class="kw">InputParamList</span>(p1),</span>
<span id="cb285-19"><a href="neoantigen-prediction.html#cb285-19"></a>                   <span class="dt">outputs =</span> <span class="kw">OutputParamList</span>(o1))</span>
<span id="cb285-20"><a href="neoantigen-prediction.html#cb285-20"></a><span class="co">## run</span></span>
<span id="cb285-21"><a href="neoantigen-prediction.html#cb285-21"></a>exps &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;abundance.tsv&quot;</span>,</span>
<span id="cb285-22"><a href="neoantigen-prediction.html#cb285-22"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb285-23"><a href="neoantigen-prediction.html#cb285-23"></a><span class="kw">names</span>(exps) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb285-24"><a href="neoantigen-prediction.html#cb285-24"></a></span>
<span id="cb285-25"><a href="neoantigen-prediction.html#cb285-25"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">kexp =</span> exps)</span>
<span id="cb285-26"><a href="neoantigen-prediction.html#cb285-26"></a>res4g &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(T2Gene, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList,</span>
<span id="cb285-27"><a href="neoantigen-prediction.html#cb285-27"></a>                     <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb285-28"><a href="neoantigen-prediction.html#cb285-28"></a>                                               <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb285-29"><a href="neoantigen-prediction.html#cb285-29"></a>                                               <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb285-30"><a href="neoantigen-prediction.html#cb285-30"></a>                                               <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb285-31"><a href="neoantigen-prediction.html#cb285-31"></a>                                                                <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb285-32"><a href="neoantigen-prediction.html#cb285-32"></a>                                                                <span class="dt">jobname =</span> <span class="st">&quot;vep&quot;</span>),</span>
<span id="cb285-33"><a href="neoantigen-prediction.html#cb285-33"></a>                                               <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb285-34"><a href="neoantigen-prediction.html#cb285-34"></a>                     <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Let’s annotate the VCFs with gene level expression values.</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="neoantigen-prediction.html#cb286-1"></a>vcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;_Ann.vcf$&quot;</span>,</span>
<span id="cb286-2"><a href="neoantigen-prediction.html#cb286-2"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb286-3"><a href="neoantigen-prediction.html#cb286-3"></a>exps &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;abundance_gene.tsv&quot;</span>,</span>
<span id="cb286-4"><a href="neoantigen-prediction.html#cb286-4"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb286-5"><a href="neoantigen-prediction.html#cb286-5"></a>ovcf &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;pt19_ANN.vcf&quot;</span>, <span class="st">&quot;pt20_ANN.vcf&quot;</span>)</span>
<span id="cb286-6"><a href="neoantigen-prediction.html#cb286-6"></a><span class="kw">names</span>(vcfs) &lt;-<span class="st"> </span><span class="kw">names</span>(exps) &lt;-<span class="st"> </span><span class="kw">names</span>(ovcf) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb286-7"><a href="neoantigen-prediction.html#cb286-7"></a></span>
<span id="cb286-8"><a href="neoantigen-prediction.html#cb286-8"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ivcf =</span> vcfs,</span>
<span id="cb286-9"><a href="neoantigen-prediction.html#cb286-9"></a>                  <span class="dt">expression =</span> exps,</span>
<span id="cb286-10"><a href="neoantigen-prediction.html#cb286-10"></a>                  <span class="dt">ovcf =</span> ovcf)</span>
<span id="cb286-11"><a href="neoantigen-prediction.html#cb286-11"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">etype =</span> <span class="st">&quot;custom&quot;</span>,</span>
<span id="cb286-12"><a href="neoantigen-prediction.html#cb286-12"></a>                  <span class="dt">gtype =</span> <span class="st">&quot;gene&quot;</span>,</span>
<span id="cb286-13"><a href="neoantigen-prediction.html#cb286-13"></a>                  <span class="dt">idCol =</span> <span class="st">&quot;gene&quot;</span>,</span>
<span id="cb286-14"><a href="neoantigen-prediction.html#cb286-14"></a>                  <span class="dt">expCol =</span> <span class="st">&quot;abundance&quot;</span>)</span>
<span id="cb286-15"><a href="neoantigen-prediction.html#cb286-15"></a></span>
<span id="cb286-16"><a href="neoantigen-prediction.html#cb286-16"></a>res5g &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(vcf_expression_annotator, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb286-17"><a href="neoantigen-prediction.html#cb286-17"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb286-18"><a href="neoantigen-prediction.html#cb286-18"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb286-19"><a href="neoantigen-prediction.html#cb286-19"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb286-20"><a href="neoantigen-prediction.html#cb286-20"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb286-21"><a href="neoantigen-prediction.html#cb286-21"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb286-22"><a href="neoantigen-prediction.html#cb286-22"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;vep&quot;</span>),</span>
<span id="cb286-23"><a href="neoantigen-prediction.html#cb286-23"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb286-24"><a href="neoantigen-prediction.html#cb286-24"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The steps were also combined into an expression annotation pipelines, <code>vcfExpression</code>. By given RNASeq raw reads, the transcript expression levels will be added to the corresponding variants in the “vep” annotated VCFs.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="neoantigen-prediction.html#cb287-1"></a><span class="kw">plotCWL</span>(vcfExpression)</span></code></pre></div>
<div id="htmlwidget-23aad275562c538b2a2c" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-23aad275562c538b2a2c">{"x":{"diagram":"graph TB\nA(rnafqs)-->F{kallistoQuant}\nB(kallistoIdx)-->F{kallistoQuant}\nC(threads)-->F{kallistoQuant}\nF{kallistoQuant}-->|tsv|G{cleanExp}\nD(svcf)-->H{vcfExpAnn}\nG{cleanExp}-->|aout|H{vcfExpAnn}\nF{kallistoQuant}-->|tsv|I{T2Gene}\nH{vcfExpAnn}-->|oVcf|J{vcfgExpAnn}\nD(svcf)-->J{vcfgExpAnn}\nI{T2Gene}-->|gout|J{vcfgExpAnn}\nJ{vcfgExpAnn}-->|oVcf|K{bgzip}\nK{bgzip}-->|zfile|L{tabixIndex}\nL{tabixIndex}-->E((ExpVcf))"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="creating-a-phased-vcf" class="section level3">
<h3><span class="header-section-number">8.1.4</span> Creating a phased VCF</h3>
<p>The germline variants and somatic variants were combined and phased in this step to calculated wildtype and mutant protein sequences more accurately. The <code>picard</code> and <code>GATK</code> tools were used to rename, combine, sort and phase the variants.</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="neoantigen-prediction.html#cb288-1"></a><span class="kw">plotCWL</span>(phaseVcf)</span></code></pre></div>
<div id="htmlwidget-3b036ebcfa738bb71fec" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-3b036ebcfa738bb71fec">{"x":{"diagram":"graph TB\nA(gvariant)-->I{splitSample}\nF(nsample)-->I{splitSample}\nI{splitSample}-->|Fout|J{renameGVcf}\nG(tsample)-->J{renameGVcf}\nB(svariant)-->K{renameSVcf}\nG(tsample)-->K{renameSVcf}\nJ{renameGVcf}-->|oVcf|L{combineVariants}\nK{renameSVcf}-->|oVcf|L{combineVariants}\nC(ref)-->L{combineVariants}\nL{combineVariants}-->|oVcf|M{sortVcf}\nM{sortVcf}-->|oVcf|N{ReadBackedPhasing}\nD(bam)-->N{ReadBackedPhasing}\nC(ref)-->N{ReadBackedPhasing}\nM{sortVcf}-->|oVcf|N{ReadBackedPhasing}\nE(outvcf)-->N{ReadBackedPhasing}\nN{ReadBackedPhasing}-->|oVcf|O{bgzip}\nO{bgzip}-->|zfile|P{tabixIndex}\nP{tabixIndex}-->H((pvcf))"},"evals":[],"jsHooks":[]}</script>
<p>To create the pahsed VCF.</p>
<div class="sourceCode" id="cb289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="neoantigen-prediction.html#cb289-1"></a>svcfs &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;apps/variants/neoantigen/pt19/pt19_Ann.vcf.gz&quot;</span>,</span>
<span id="cb289-2"><a href="neoantigen-prediction.html#cb289-2"></a>              <span class="dt">pt20 =</span> <span class="st">&quot;apps/variants/neoantigen/pt20/pt20_Ann.vcf.gz&quot;</span>)</span>
<span id="cb289-3"><a href="neoantigen-prediction.html#cb289-3"></a>gvcfs &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;apps/variants/vcf/res/output/germline.vcf.gz&quot;</span>,</span>
<span id="cb289-4"><a href="neoantigen-prediction.html#cb289-4"></a>              <span class="dt">pt20 =</span> <span class="st">&quot;apps/variants/vcf/res/output/germline.vcf.gz&quot;</span>)</span>
<span id="cb289-5"><a href="neoantigen-prediction.html#cb289-5"></a>bam &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;apps/variants/BAM/tumor19/tumor19.bam&quot;</span>,</span>
<span id="cb289-6"><a href="neoantigen-prediction.html#cb289-6"></a>            <span class="dt">pt20 =</span> <span class="st">&quot;apps/variants/BAM/tumor20/tumor20.bam&quot;</span>)</span>
<span id="cb289-7"><a href="neoantigen-prediction.html#cb289-7"></a>outvcf &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;pt19_phased.vcf&quot;</span>,</span>
<span id="cb289-8"><a href="neoantigen-prediction.html#cb289-8"></a>               <span class="dt">pt20 =</span> <span class="st">&quot;pt20_phased.vcf&quot;</span>)</span>
<span id="cb289-9"><a href="neoantigen-prediction.html#cb289-9"></a>nsamples &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;normal19&quot;</span>, <span class="st">&quot;normal20&quot;</span>)</span>
<span id="cb289-10"><a href="neoantigen-prediction.html#cb289-10"></a>tsamples &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;tumor19&quot;</span>, <span class="st">&quot;tumor20&quot;</span>)</span>
<span id="cb289-11"><a href="neoantigen-prediction.html#cb289-11"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">gvariant =</span> gvcfs,</span>
<span id="cb289-12"><a href="neoantigen-prediction.html#cb289-12"></a>                  <span class="dt">svariant =</span> svcfs,</span>
<span id="cb289-13"><a href="neoantigen-prediction.html#cb289-13"></a>                  <span class="dt">bam =</span> bam,</span>
<span id="cb289-14"><a href="neoantigen-prediction.html#cb289-14"></a>                  <span class="dt">outvcf =</span> outvcf,</span>
<span id="cb289-15"><a href="neoantigen-prediction.html#cb289-15"></a>                  <span class="dt">nsample =</span> nsamples,</span>
<span id="cb289-16"><a href="neoantigen-prediction.html#cb289-16"></a>                  <span class="dt">tsample =</span> tsamples)</span>
<span id="cb289-17"><a href="neoantigen-prediction.html#cb289-17"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ref =</span> ref)</span>
<span id="cb289-18"><a href="neoantigen-prediction.html#cb289-18"></a></span>
<span id="cb289-19"><a href="neoantigen-prediction.html#cb289-19"></a>res6 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(phaseVcf, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb289-20"><a href="neoantigen-prediction.html#cb289-20"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb289-21"><a href="neoantigen-prediction.html#cb289-21"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb289-22"><a href="neoantigen-prediction.html#cb289-22"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb289-23"><a href="neoantigen-prediction.html#cb289-23"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb289-24"><a href="neoantigen-prediction.html#cb289-24"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb289-25"><a href="neoantigen-prediction.html#cb289-25"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;phase&quot;</span>),</span>
<span id="cb289-26"><a href="neoantigen-prediction.html#cb289-26"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb289-27"><a href="neoantigen-prediction.html#cb289-27"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The phased VCF file using HP tags to link alleles. Here are the phased VCFs.</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="neoantigen-prediction.html#cb290-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>, <span class="dt">glob =</span> <span class="st">&quot;*phased.vcf*&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 4 x 2
##   path                                                           size
##   &lt;fs::path&gt;                                              &lt;fs::bytes&gt;
## 1 apps/variants/neoantigen/pt19/tumor19_phased.vcf.gz           4.84M
## 2 apps/variants/neoantigen/pt19/tumor19_phased.vcf.gz.tbi      31.92K
## 3 apps/variants/neoantigen/pt20/tumor20_phased.vcf.gz            4.3M
## 4 apps/variants/neoantigen/pt20/tumor20_phased.vcf.gz.tbi      34.06K</code></pre>
</div>
<div id="combined-vcf-preparing-pipeline" class="section level3">
<h3><span class="header-section-number">8.1.5</span> Combined VCF preparing pipeline</h3>
<p>The previous 4 annotation steps have been intergrated into one VCF annotation pipeline. Thus all the required inputs can be prepared in just one step.</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="neoantigen-prediction.html#cb292-1"></a><span class="kw">plotCWL</span>(AnnPhaseVcf)</span></code></pre></div>
<div id="htmlwidget-a6d192b0dbdd3ead48be" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-a6d192b0dbdd3ead48be">{"x":{"diagram":"graph TB\nA(svcf)-->N{VCFvep}\nC(ref)-->N{VCFvep}\nD(VepDir)-->N{VCFvep}\nN{VCFvep}-->|oVcf|O{dVCFcoverage}\nE(tbam)-->O{dVCFcoverage}\nC(ref)-->O{dVCFcoverage}\nO{dVCFcoverage}-->|outvcf|P{rVCFcoverage}\nF(rbam)-->P{rVCFcoverage}\nC(ref)-->P{rVCFcoverage}\nI(rnaseqs)-->Q{VCFexpression}\nJ(kallistoIdx)-->Q{VCFexpression}\nP{rVCFcoverage}-->|outvcf|Q{VCFexpression}\nK(threads)-->Q{VCFexpression}\nQ{VCFexpression}-->L((annVcf))\nB(gvcf)-->R{PhaseVcf}\nL((annVcf))-->R{PhaseVcf}\nE(tbam)-->R{PhaseVcf}\nH(nsample)-->R{PhaseVcf}\nG(tsample)-->R{PhaseVcf}\nC(ref)-->R{PhaseVcf}\nR{PhaseVcf}-->M((phasedVCF))"},"evals":[],"jsHooks":[]}</script>
<p>Here are the example to run the annotation and phasing pipeline.</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="neoantigen-prediction.html#cb293-1"></a>svcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/combined&quot;</span>, <span class="st">&quot;.neusomatic.vcf&quot;</span>,</span>
<span id="cb293-2"><a href="neoantigen-prediction.html#cb293-2"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb293-3"><a href="neoantigen-prediction.html#cb293-3"></a><span class="kw">names</span>(svcfs) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb293-4"><a href="neoantigen-prediction.html#cb293-4"></a>gvcfs &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;apps/variants/vcf/res/output/germline.vcf.gz&quot;</span>,</span>
<span id="cb293-5"><a href="neoantigen-prediction.html#cb293-5"></a>              <span class="dt">pt20 =</span> <span class="st">&quot;apps/variants/vcf/res/output/germline.vcf.gz&quot;</span>)</span>
<span id="cb293-6"><a href="neoantigen-prediction.html#cb293-6"></a>bam &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;apps/variants/BAM/tumor19/tumor19.bam&quot;</span>,</span>
<span id="cb293-7"><a href="neoantigen-prediction.html#cb293-7"></a>            <span class="dt">pt20 =</span> <span class="st">&quot;apps/variants/BAM/tumor20/tumor20.bam&quot;</span>)</span>
<span id="cb293-8"><a href="neoantigen-prediction.html#cb293-8"></a><span class="co">## use rnaseq BAM instead</span></span>
<span id="cb293-9"><a href="neoantigen-prediction.html#cb293-9"></a>rbam &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="st">&quot;apps/variants/BAM/tumor19/tumor19.bam&quot;</span>,</span>
<span id="cb293-10"><a href="neoantigen-prediction.html#cb293-10"></a>             <span class="dt">pt20 =</span> <span class="st">&quot;apps/variants/BAM/tumor20/tumor20.bam&quot;</span>)</span>
<span id="cb293-11"><a href="neoantigen-prediction.html#cb293-11"></a>fqs &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&quot;apps/variants/fastq&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb293-12"><a href="neoantigen-prediction.html#cb293-12"></a>fqs &lt;-<span class="st"> </span>fqs[<span class="kw">grep</span>(<span class="st">&quot;tumor&quot;</span>, fqs)]</span>
<span id="cb293-13"><a href="neoantigen-prediction.html#cb293-13"></a>fqL &lt;-<span class="st"> </span><span class="kw">tapply</span>(fqs, <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>), <span class="dt">each =</span> <span class="dv">2</span>), as.list)</span>
<span id="cb293-14"><a href="neoantigen-prediction.html#cb293-14"></a>nsamples &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;normal19&quot;</span>, <span class="st">&quot;normal20&quot;</span>)</span>
<span id="cb293-15"><a href="neoantigen-prediction.html#cb293-15"></a>tsamples &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;tumor19&quot;</span>, <span class="st">&quot;tumor20&quot;</span>)</span>
<span id="cb293-16"><a href="neoantigen-prediction.html#cb293-16"></a></span>
<span id="cb293-17"><a href="neoantigen-prediction.html#cb293-17"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">gvcf =</span> gvcfs,</span>
<span id="cb293-18"><a href="neoantigen-prediction.html#cb293-18"></a>                  <span class="dt">svcf =</span> svcfs,</span>
<span id="cb293-19"><a href="neoantigen-prediction.html#cb293-19"></a>                  <span class="dt">tbam =</span> bam,</span>
<span id="cb293-20"><a href="neoantigen-prediction.html#cb293-20"></a>                  <span class="dt">nsample =</span> nsamples,</span>
<span id="cb293-21"><a href="neoantigen-prediction.html#cb293-21"></a>                  <span class="dt">tsample =</span> tsamples,</span>
<span id="cb293-22"><a href="neoantigen-prediction.html#cb293-22"></a>                  <span class="dt">rnaseqs =</span> fqL,</span>
<span id="cb293-23"><a href="neoantigen-prediction.html#cb293-23"></a>                  <span class="dt">rbam =</span> rbam)</span>
<span id="cb293-24"><a href="neoantigen-prediction.html#cb293-24"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ref =</span> ref,</span>
<span id="cb293-25"><a href="neoantigen-prediction.html#cb293-25"></a>                  <span class="dt">VepDir =</span> <span class="st">&quot;~/.vep/&quot;</span>,</span>
<span id="cb293-26"><a href="neoantigen-prediction.html#cb293-26"></a>                  <span class="dt">kallistoIdx =</span> <span class="st">&quot;apps/data/gencode.v25.transcripts_kallisto&quot;</span>,</span>
<span id="cb293-27"><a href="neoantigen-prediction.html#cb293-27"></a>                  <span class="dt">threads =</span> <span class="dv">16</span>)</span>
<span id="cb293-28"><a href="neoantigen-prediction.html#cb293-28"></a></span>
<span id="cb293-29"><a href="neoantigen-prediction.html#cb293-29"></a>res7 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(AnnPhaseVcf, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb293-30"><a href="neoantigen-prediction.html#cb293-30"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb293-31"><a href="neoantigen-prediction.html#cb293-31"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb293-32"><a href="neoantigen-prediction.html#cb293-32"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb293-33"><a href="neoantigen-prediction.html#cb293-33"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb293-34"><a href="neoantigen-prediction.html#cb293-34"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb293-35"><a href="neoantigen-prediction.html#cb293-35"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;AnnPhase&quot;</span>),</span>
<span id="cb293-36"><a href="neoantigen-prediction.html#cb293-36"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb293-37"><a href="neoantigen-prediction.html#cb293-37"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Here are the results:</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="neoantigen-prediction.html#cb294-1"></a><span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen/pt19&quot;</span>, <span class="st">&quot;vcf.gz&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;pt19_Ann.vcf.gz&quot;                                                        
## [2] &quot;pt19_Ann.vcf.gz.tbi&quot;                                                    
## [3] &quot;tumor19_neusomatic_vep_dc_snv_indel_dc_snv_indel_ExpAnn_gAnn.vcf.gz&quot;    
## [4] &quot;tumor19_neusomatic_vep_dc_snv_indel_dc_snv_indel_ExpAnn_gAnn.vcf.gz.tbi&quot;
## [5] &quot;tumor19_neusomatic_vep_dc_snv_indel_ExpAnn.vcf.gz&quot;                      
## [6] &quot;tumor19_neusomatic_vep_dc_snv_indel_ExpAnn.vcf.gz.tbi&quot;                  
## [7] &quot;tumor19_phased.vcf.gz&quot;                                                  
## [8] &quot;tumor19_phased.vcf.gz.tbi&quot;</code></pre>
</div>
</div>
<div id="hla-typing" class="section level2">
<h2><span class="header-section-number">8.2</span> HLA typing</h2>
<div id="polysolver" class="section level3">
<h3><span class="header-section-number">8.2.1</span> polysolver</h3>
<p>The <code>polysolver</code> is a tool for HLA typing based on whole exome sequencing data. The details to run the tool can be found: <a href="https://software.broadinstitute.org/cancer/cga/polysolver_run" class="uri">https://software.broadinstitute.org/cancer/cga/polysolver_run</a>.</p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="neoantigen-prediction.html#cb296-1"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">bam =</span> bam)</span>
<span id="cb296-2"><a href="neoantigen-prediction.html#cb296-2"></a>res8 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(polysolver, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList,</span>
<span id="cb296-3"><a href="neoantigen-prediction.html#cb296-3"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb296-4"><a href="neoantigen-prediction.html#cb296-4"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb296-5"><a href="neoantigen-prediction.html#cb296-5"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb296-6"><a href="neoantigen-prediction.html#cb296-6"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb296-7"><a href="neoantigen-prediction.html#cb296-7"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb296-8"><a href="neoantigen-prediction.html#cb296-8"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;polysolver&quot;</span>),</span>
<span id="cb296-9"><a href="neoantigen-prediction.html#cb296-9"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb296-10"><a href="neoantigen-prediction.html#cb296-10"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>, <span class="dt">cwlTemp =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
</div>
<div id="optitype" class="section level3">
<h3><span class="header-section-number">8.2.2</span> Optitype</h3>
</div>
</div>
<div id="neoantigen-prediction-by-pvacseq" class="section level2">
<h2><span class="header-section-number">8.3</span> Neoantigen prediction by pVACseq</h2>
<p>The <code>pvacseq</code> tool take the annotated somatic variants and phased combined variants as inputs. Multiple candidate alleles from hla typing and multiple epitope prediction algorithms can be used.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="neoantigen-prediction.html#cb297-1"></a>vcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;gAnn.vcf.gz$&quot;</span>,</span>
<span id="cb297-2"><a href="neoantigen-prediction.html#cb297-2"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb297-3"><a href="neoantigen-prediction.html#cb297-3"></a>pvcfs &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen&quot;</span>, <span class="st">&quot;phased.vcf.gz$&quot;</span>,</span>
<span id="cb297-4"><a href="neoantigen-prediction.html#cb297-4"></a>                           <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb297-5"><a href="neoantigen-prediction.html#cb297-5"></a><span class="kw">names</span>(vcfs) &lt;-<span class="st"> </span><span class="kw">names</span>(pvcfs) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb297-6"><a href="neoantigen-prediction.html#cb297-6"></a>alleles &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pt19 =</span> <span class="kw">list</span>(<span class="st">&quot;HLA-A*02:01&quot;</span>, <span class="st">&quot;HLA-B*35:01&quot;</span>),</span>
<span id="cb297-7"><a href="neoantigen-prediction.html#cb297-7"></a>                <span class="dt">pt20 =</span> <span class="kw">list</span>(<span class="st">&quot;HLA-A*02:01&quot;</span>, <span class="st">&quot;HLA-B*35:01&quot;</span>, <span class="st">&quot;DRB1*11:01&quot;</span>))</span>
<span id="cb297-8"><a href="neoantigen-prediction.html#cb297-8"></a></span>
<span id="cb297-9"><a href="neoantigen-prediction.html#cb297-9"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ivcf =</span> vcfs,</span>
<span id="cb297-10"><a href="neoantigen-prediction.html#cb297-10"></a>                  <span class="dt">phasedVcf =</span> pvcfs,</span>
<span id="cb297-11"><a href="neoantigen-prediction.html#cb297-11"></a>                  <span class="dt">sample =</span> tsamples,</span>
<span id="cb297-12"><a href="neoantigen-prediction.html#cb297-12"></a>                  <span class="dt">allele =</span> alleles)</span>
<span id="cb297-13"><a href="neoantigen-prediction.html#cb297-13"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">algorithms =</span> <span class="kw">list</span>(<span class="st">&quot;MHCflurry&quot;</span>, <span class="st">&quot;MHCnuggetsI&quot;</span>,</span>
<span id="cb297-14"><a href="neoantigen-prediction.html#cb297-14"></a>                                    <span class="st">&quot;MHCnuggetsII&quot;</span>, <span class="st">&quot;NNalign&quot;</span>, <span class="st">&quot;NetMHC&quot;</span>))</span>
<span id="cb297-15"><a href="neoantigen-prediction.html#cb297-15"></a></span>
<span id="cb297-16"><a href="neoantigen-prediction.html#cb297-16"></a>res9 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(pvacseq, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/neoantigen&quot;</span>, inputList, paramList,</span>
<span id="cb297-17"><a href="neoantigen-prediction.html#cb297-17"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb297-18"><a href="neoantigen-prediction.html#cb297-18"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb297-19"><a href="neoantigen-prediction.html#cb297-19"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb297-20"><a href="neoantigen-prediction.html#cb297-20"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb297-21"><a href="neoantigen-prediction.html#cb297-21"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb297-22"><a href="neoantigen-prediction.html#cb297-22"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;pvacseq&quot;</span>),</span>
<span id="cb297-23"><a href="neoantigen-prediction.html#cb297-23"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb297-24"><a href="neoantigen-prediction.html#cb297-24"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Here are the filtered and ranked results:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="neoantigen-prediction.html#cb298-1"></a><span class="kw">list.files</span>(<span class="st">&quot;apps/variants/neoantigen/pt20/pvacseq_out/combined/&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;tumor20.all_epitopes.tsv&quot;             
## [2] &quot;tumor20.filtered.condensed.ranked.tsv&quot;
## [3] &quot;tumor20.filtered.tsv&quot;</code></pre>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="neoantigen-prediction.html#cb300-1"></a>neo &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;apps/variants/neoantigen/pt20/pvacseq_out/combined/tumor20.filtered.condensed.ranked.tsv&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">check.names =</span> <span class="ot">FALSE</span>)</span>
<span id="cb300-2"><a href="neoantigen-prediction.html#cb300-2"></a><span class="kw">head</span>(neo)</span></code></pre></div>
<pre><code>##   Gene Name Mutation Protein Position                       HGVSc
## 1      GNAS      K/M              876 ENST00000371100.4:c.2627A&gt;T
## 2   MACROD2      G/A               59  ENST00000407045.3:c.176G&gt;C
## 3     PLCG1      R/H              886 ENST00000373271.1:c.2657G&gt;A
## 4      RIN2      V/G              564 ENST00000255006.6:c.1691T&gt;G
## 5   SLC4A11      F/S              647 ENST00000380059.3:c.1940T&gt;C
## 6   SLC4A11      S/P              562 ENST00000474451.1:c.1684T&gt;C
##                           HGVSp  HLA Allele Mutation Position
## 1 ENSP00000360141.3:p.Lys876Met HLA-A*02:01                 2
## 2  ENSP00000385516.3:p.Gly59Ala HLA-B*35:01                 3
## 3 ENSP00000362368.1:p.Arg886His HLA-B*35:01                 9
## 4 ENSP00000255006.6:p.Val564Gly  DRB1*11:01                10
## 5 ENSP00000369399.3:p.Phe647Ser HLA-B*35:01                 8
## 6 ENSP00000476859.1:p.Ser562Pro HLA-B*35:01                 3
##    MT Epitope Seq Median MT Score Median WT Score Median Fold Change
## 1      RMWIQCFNDV          69.936        9047.261            129.364
## 2        TPAPDVEM          47.965         276.192              5.758
## 3       VPACQIAIH         346.080        3384.033              9.778
## 4 SFMTPEKRMGRRIAE          30.461          19.712              0.647
## 5      LPIAVLASSL         114.193          55.791              0.489
## 6       SPPPSSAPM         111.499          11.532              0.103
##   Best MT Score Corresponding WT Score Corresponding Fold Change
## 1        42.920               7634.610                   177.880
## 2        39.085                276.192                     7.066
## 3       225.113                724.975                     3.220
## 4        13.200                 12.400                     0.939
## 5        16.548                  7.875                     0.476
## 6        52.980                  9.080                     0.171
##   Tumor DNA Depth Tumor DNA VAF Tumor RNA Depth Tumor RNA VAF
## 1              32         0.344              32         0.344
## 2              16         0.375              16         0.375
## 3              41         0.390              41         0.390
## 4              28         0.393              28         0.393
## 5              42         0.357              42         0.357
## 6              42         0.357              42         0.357
##   Gene Expression Rank
## 1        1922.752    1
## 2        1287.893    2
## 3        2025.111    3
## 4        1153.660    4
## 5         935.994    5
## 6         935.994    6</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="rna-seq-alignment-and-quantification.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-sequencing-pipelines.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/08-Neoantigen.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
