<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 RNA-Seq alignment and quantification | Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 RNA-Seq alignment and quantification | Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 RNA-Seq alignment and quantification | Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2019-10-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="dna-seq-variant-calling.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/d3-3.3.8/d3.min.js"></script>
<script src="libs/dagre-0.4.0/dagre-d3.min.js"></script>
<link href="libs/mermaid-0.3.0/dist/mermaid.css" rel="stylesheet" />
<script src="libs/mermaid-0.3.0/dist/mermaid.slim.min.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/chromatography-0.1/chromatography.js"></script>
<script src="libs/DiagrammeR-binding-1.0.1/DiagrammeR.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#common-workflow-language"><i class="fa fa-check"></i><b>1.1</b> Common Workflow Language</a></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#first-example"><i class="fa fa-check"></i><b>1.2</b> First example</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#test-run"><i class="fa fa-check"></i><b>1.3</b> Test run</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="components.html"><a href="components.html"><i class="fa fa-check"></i><b>2</b> Components</a><ul>
<li class="chapter" data-level="2.1" data-path="components.html"><a href="components.html#input-parameters"><i class="fa fa-check"></i><b>2.1</b> Input Parameters</a><ul>
<li class="chapter" data-level="2.1.1" data-path="components.html"><a href="components.html#essential-input-parameters"><i class="fa fa-check"></i><b>2.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="2.1.2" data-path="components.html"><a href="components.html#array-inputs"><i class="fa fa-check"></i><b>2.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="components.html"><a href="components.html#output-parameters"><i class="fa fa-check"></i><b>2.2</b> Output Parameters</a><ul>
<li class="chapter" data-level="2.2.1" data-path="components.html"><a href="components.html#capturing-output"><i class="fa fa-check"></i><b>2.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="2.2.2" data-path="components.html"><a href="components.html#array-outputs"><i class="fa fa-check"></i><b>2.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="2.2.3" data-path="components.html"><a href="components.html#standard-output"><i class="fa fa-check"></i><b>2.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>3</b> Writing Pipeline</a><ul>
<li class="chapter" data-level="3.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>3.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="3.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>3.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="run-approaches.html"><a href="run-approaches.html"><i class="fa fa-check"></i><b>4</b> Run approaches</a><ul>
<li class="chapter" data-level="4.1" data-path="run-approaches.html"><a href="run-approaches.html#running-tools-in-docker"><i class="fa fa-check"></i><b>4.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="4.2" data-path="run-approaches.html"><a href="run-approaches.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>4.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="4.3" data-path="run-approaches.html"><a href="run-approaches.html#web-application"><i class="fa fa-check"></i><b>4.3</b> Web Application</a><ul>
<li class="chapter" data-level="4.3.1" data-path="run-approaches.html"><a href="run-approaches.html#cwlparam-example"><i class="fa fa-check"></i><b>4.3.1</b> cwlParam example</a></li>
<li class="chapter" data-level="4.3.2" data-path="run-approaches.html"><a href="run-approaches.html#cwlparam-to-shiny-app"><i class="fa fa-check"></i><b>4.3.2</b> cwlParam to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="application.html"><a href="application.html"><i class="fa fa-check"></i><b>5</b> Application</a><ul>
<li class="chapter" data-level="5.1" data-path="application.html"><a href="application.html#rcwlpipelines-tools"><i class="fa fa-check"></i><b>5.1</b> RcwlPipelines tools</a><ul>
<li class="chapter" data-level="5.1.1" data-path="application.html"><a href="application.html#rcwl-scripts"><i class="fa fa-check"></i><b>5.1.1</b> Rcwl scripts</a></li>
<li class="chapter" data-level="5.1.2" data-path="application.html"><a href="application.html#build-a-pipeline"><i class="fa fa-check"></i><b>5.1.2</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="application.html"><a href="application.html#rcwlpipelines-summary"><i class="fa fa-check"></i><b>5.2</b> RcwlPipelines summary</a></li>
<li class="chapter" data-level="5.3" data-path="application.html"><a href="application.html#dnaseq-alignment-pipeline"><i class="fa fa-check"></i><b>5.3</b> DNASeq alignment pipeline</a><ul>
<li class="chapter" data-level="5.3.1" data-path="application.html"><a href="application.html#prepare-data"><i class="fa fa-check"></i><b>5.3.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.3.2" data-path="application.html"><a href="application.html#run-in-cluster"><i class="fa fa-check"></i><b>5.3.2</b> Run in cluster</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="application.html"><a href="application.html#rnaseq-pipeline"><i class="fa fa-check"></i><b>5.4</b> RNASeq pipeline</a><ul>
<li class="chapter" data-level="5.4.1" data-path="application.html"><a href="application.html#prepare-data-1"><i class="fa fa-check"></i><b>5.4.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.4.2" data-path="application.html"><a href="application.html#submit-pipeline-with-sge"><i class="fa fa-check"></i><b>5.4.2</b> Submit pipeline with SGE</a></li>
<li class="chapter" data-level="5.4.3" data-path="application.html"><a href="application.html#summarize-qc"><i class="fa fa-check"></i><b>5.4.3</b> Summarize QC</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="application.html"><a href="application.html#mc3-somatic-variant-calling-pipeline"><i class="fa fa-check"></i><b>5.5</b> MC3 somatic variant calling pipeline</a><ul>
<li class="chapter" data-level="5.5.1" data-path="application.html"><a href="application.html#prepare-data-2"><i class="fa fa-check"></i><b>5.5.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.5.2" data-path="application.html"><a href="application.html#run-mc3-pipeline"><i class="fa fa-check"></i><b>5.5.2</b> Run MC3 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="application.html"><a href="application.html#gatk4-germline-variant-calling-pipeline"><i class="fa fa-check"></i><b>5.6</b> GATK4 germline variant calling pipeline</a><ul>
<li class="chapter" data-level="5.6.1" data-path="application.html"><a href="application.html#gatk-alignment"><i class="fa fa-check"></i><b>5.6.1</b> GATK Alignment</a></li>
<li class="chapter" data-level="5.6.2" data-path="application.html"><a href="application.html#haplotypecaller"><i class="fa fa-check"></i><b>5.6.2</b> HaplotypeCaller</a></li>
<li class="chapter" data-level="5.6.3" data-path="application.html"><a href="application.html#joint-discovery"><i class="fa fa-check"></i><b>5.6.3</b> Joint Discovery</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html"><i class="fa fa-check"></i><b>6</b> DNA-Seq Variant calling</a><ul>
<li class="chapter" data-level="6.1" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#dna-alignment"><i class="fa fa-check"></i><b>6.1</b> DNA alignment</a></li>
<li class="chapter" data-level="6.2" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#germline-variant-calling"><i class="fa fa-check"></i><b>6.2</b> Germline variant calling</a></li>
<li class="chapter" data-level="6.3" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#somatic-mutation-calling"><i class="fa fa-check"></i><b>6.3</b> Somatic mutation calling</a><ul>
<li class="chapter" data-level="6.3.1" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#mutect2"><i class="fa fa-check"></i><b>6.3.1</b> MuTect2</a></li>
<li class="chapter" data-level="6.3.2" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#muse"><i class="fa fa-check"></i><b>6.3.2</b> MuSE</a></li>
<li class="chapter" data-level="6.3.3" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#strelka2"><i class="fa fa-check"></i><b>6.3.3</b> Strelka2</a></li>
<li class="chapter" data-level="6.3.4" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#somaticsniper"><i class="fa fa-check"></i><b>6.3.4</b> SomaticSniper</a></li>
<li class="chapter" data-level="6.3.5" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#vardict"><i class="fa fa-check"></i><b>6.3.5</b> VarDict</a></li>
<li class="chapter" data-level="6.3.6" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#lofreq"><i class="fa fa-check"></i><b>6.3.6</b> LoFreq</a></li>
<li class="chapter" data-level="6.3.7" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#varscan2"><i class="fa fa-check"></i><b>6.3.7</b> VarScan2</a></li>
<li class="chapter" data-level="6.3.8" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#lancet"><i class="fa fa-check"></i><b>6.3.8</b> Lancet</a></li>
<li class="chapter" data-level="6.3.9" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#neusomatic"><i class="fa fa-check"></i><b>6.3.9</b> Neusomatic</a></li>
<li class="chapter" data-level="6.3.10" data-path="dna-seq-variant-calling.html"><a href="dna-seq-variant-calling.html#combined-somatic-caller-pipeline"><i class="fa fa-check"></i><b>6.3.10</b> Combined somatic caller pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html"><i class="fa fa-check"></i><b>7</b> RNA-Seq alignment and quantification</a><ul>
<li class="chapter" data-level="7.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#bulk-mrna"><i class="fa fa-check"></i><b>7.1</b> bulk mRNA</a></li>
<li class="chapter" data-level="7.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#transcriptome-quantification"><i class="fa fa-check"></i><b>7.2</b> transcriptome quantification</a></li>
<li class="chapter" data-level="7.3" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#single-cell-rnaseq"><i class="fa fa-check"></i><b>7.3</b> single cell RNAseq</a></li>
<li class="chapter" data-level="7.4" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#mirna"><i class="fa fa-check"></i><b>7.4</b> miRNA</a><ul>
<li class="chapter" data-level="7.4.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#prepare-reference"><i class="fa fa-check"></i><b>7.4.1</b> Prepare reference</a></li>
<li class="chapter" data-level="7.4.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>7.4.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rna-seq-alignment-and-quantification" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> RNA-Seq alignment and quantification</h1>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="rna-seq-alignment-and-quantification.html#cb218-1"></a><span class="kw">library</span>(RcwlPipelines)</span>
<span id="cb218-2"><a href="rna-seq-alignment-and-quantification.html#cb218-2"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb218-3"><a href="rna-seq-alignment-and-quantification.html#cb218-3"></a><span class="kw">library</span>(fs)</span></code></pre></div>
<div id="bulk-mrna" class="section level2">
<h2><span class="header-section-number">7.1</span> bulk mRNA</h2>
</div>
<div id="transcriptome-quantification" class="section level2">
<h2><span class="header-section-number">7.2</span> transcriptome quantification</h2>
</div>
<div id="single-cell-rnaseq" class="section level2">
<h2><span class="header-section-number">7.3</span> single cell RNAseq</h2>
</div>
<div id="mirna" class="section level2">
<h2><span class="header-section-number">7.4</span> miRNA</h2>
<p>The <code>miRDeep2</code> is one of the popular tools for discovering known and novel miRNAs from small RNA sequencing data. We wrapped the mapping and quantification steps into <code>miRDeep2PL</code>. Here is the instructions for <code>miRDeep2</code>: <a href="https://github.com/rajewsky-lab/mirdeep2" class="uri">https://github.com/rajewsky-lab/mirdeep2</a>.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="rna-seq-alignment-and-quantification.html#cb219-1"></a><span class="kw">plotCWL</span>(miRDeep2PL)</span></code></pre></div>
<div id="htmlwidget-958121800c921e741fbe" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-958121800c921e741fbe">{"x":{"diagram":"graph TB\nA(reads)-->R{Mapper}\nB(format)-->R{Mapper}\nC(adapter)-->R{Mapper}\nE(genome)-->R{Mapper}\nR{Mapper}-->P((preads))\nR{Mapper}-->Q((arf))\nP((preads))-->S{miRDeep2}\nE(genome)-->S{miRDeep2}\nQ((arf))-->S{miRDeep2}\nF(miRef)-->S{miRDeep2}\nG(miOther)-->S{miRDeep2}\nH(precursors)-->S{miRDeep2}\nI(species)-->S{miRDeep2}\nS{miRDeep2}-->J((csvfiles))\nS{miRDeep2}-->K((htmls))\nS{miRDeep2}-->L((bed))\nS{miRDeep2}-->M((expression))\nS{miRDeep2}-->N((mirna_results))\nS{miRDeep2}-->O((pdfs))"},"evals":[],"jsHooks":[]}</script>
<p>We use the data from <code>mirdeep2</code> github repository as an example.
<a href="https://github.com/rajewsky-lab/mirdeep2/tree/master/tutorial_dir" class="uri">https://github.com/rajewsky-lab/mirdeep2/tree/master/tutorial_dir</a></p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="rna-seq-alignment-and-quantification.html#cb220-1"></a><span class="kw">list.files</span>(<span class="st">&quot;apps/miRNA/tutorial_dir&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;cel_cluster.fa&quot;                 &quot;mature_ref_other_species.fa&quot;   
## [3] &quot;mature_ref_this_species.fa&quot;     &quot;precursors_ref_this_species.fa&quot;
## [5] &quot;reads_collapsed.fa&quot;             &quot;reads.fa&quot;                      
## [7] &quot;run_tut.sh&quot;                     &quot;sample_result.html&quot;</code></pre>
<div id="prepare-reference" class="section level3">
<h3><span class="header-section-number">7.4.1</span> Prepare reference</h3>
<p>First, we need to build indexes for the miRNA reference with <code>bowtie-build</code>. This is only required to be performed once for each refernce genome.</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="rna-seq-alignment-and-quantification.html#cb222-1"></a>bowtie_build<span class="op">$</span>ref &lt;-<span class="st"> &quot;apps/miRNA/tutorial_dir/cel_cluster.fa&quot;</span></span>
<span id="cb222-2"><a href="rna-seq-alignment-and-quantification.html#cb222-2"></a>bowtie_build<span class="op">$</span>outPrefix &lt;-<span class="st"> &quot;cel_cluster&quot;</span></span>
<span id="cb222-3"><a href="rna-seq-alignment-and-quantification.html#cb222-3"></a>idxRes &lt;-<span class="st"> </span><span class="kw">runCWL</span>(bowtie_build, <span class="dt">outdir =</span> <span class="st">&quot;apps/miRNA/output/genome&quot;</span>, <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb222-4"><a href="rna-seq-alignment-and-quantification.html#cb222-4"></a><span class="kw">file.copy</span>(<span class="st">&quot;apps/miRNA/tutorial_dir/cel_cluster.fa&quot;</span>, <span class="st">&quot;apps/miRNA/output/genome/cel_cluster.fa&quot;</span>)</span></code></pre></div>
<p>Here are the indexed reference files.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="rna-seq-alignment-and-quantification.html#cb223-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/miRNA/output/genome&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 7 x 2
##   path                                                   size
##   &lt;fs::path&gt;                                      &lt;fs::bytes&gt;
## 1 apps/miRNA/output/genome/cel_cluster.1.ebwt              4M
## 2 apps/miRNA/output/genome/cel_cluster.2.ebwt             812
## 3 apps/miRNA/output/genome/cel_cluster.3.ebwt              26
## 4 apps/miRNA/output/genome/cel_cluster.4.ebwt           1.57K
## 5 apps/miRNA/output/genome/cel_cluster.fa               6.47K
## 6 apps/miRNA/output/genome/cel_cluster.rev.1.ebwt          4M
## 7 apps/miRNA/output/genome/cel_cluster.rev.2.ebwt         812</code></pre>
</div>
<div id="run-mirdeep2-pipeline" class="section level3">
<h3><span class="header-section-number">7.4.2</span> Run miRDeep2 pipeline</h3>
<p>First we need to prepare the inputs into ‘inputList’ and ‘paramList’ for parallel execution.</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="rna-seq-alignment-and-quantification.html#cb225-1"></a><span class="kw">inputs</span>(miRDeep2PL)</span></code></pre></div>
<pre><code>## inputs:
##   reads (File):  
##   format (string):  -c
##   adapter (string):  
##   len (int):  18
##   genome (File):  
##   miRef  ( File|string ): 
##   miOther  ( File|string ): 
##   precursors  ( File|string ): 
##   species (string):</code></pre>
<p>To mimic multiple samples analysis, here we just repeat to use the input reads.</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="rna-seq-alignment-and-quantification.html#cb227-1"></a>reads &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sample1 =</span> <span class="st">&quot;apps/miRNA/tutorial_dir/reads.fa&quot;</span>,</span>
<span id="cb227-2"><a href="rna-seq-alignment-and-quantification.html#cb227-2"></a>              <span class="dt">sample2 =</span> <span class="st">&quot;apps/miRNA/tutorial_dir/reads.fa&quot;</span>)</span>
<span id="cb227-3"><a href="rna-seq-alignment-and-quantification.html#cb227-3"></a></span>
<span id="cb227-4"><a href="rna-seq-alignment-and-quantification.html#cb227-4"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">reads =</span> reads)</span>
<span id="cb227-5"><a href="rna-seq-alignment-and-quantification.html#cb227-5"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">adapter =</span> <span class="st">&quot;TCGTATGCCGTCTTCTGCTTGT&quot;</span>,</span>
<span id="cb227-6"><a href="rna-seq-alignment-and-quantification.html#cb227-6"></a>                  <span class="dt">genome =</span> <span class="st">&quot;apps/miRNA/output/genome/cel_cluster.fa&quot;</span>,</span>
<span id="cb227-7"><a href="rna-seq-alignment-and-quantification.html#cb227-7"></a>                  <span class="dt">miRef =</span> <span class="st">&quot;apps/miRNA/tutorial_dir/mature_ref_this_species.fa&quot;</span>,</span>
<span id="cb227-8"><a href="rna-seq-alignment-and-quantification.html#cb227-8"></a>                  <span class="dt">miOther =</span> <span class="st">&quot;apps/miRNA/tutorial_dir/mature_ref_other_species.fa&quot;</span>,</span>
<span id="cb227-9"><a href="rna-seq-alignment-and-quantification.html#cb227-9"></a>                  <span class="dt">precursors =</span> <span class="st">&quot;apps/miRNA/tutorial_dir/precursors_ref_this_species.fa&quot;</span>,</span>
<span id="cb227-10"><a href="rna-seq-alignment-and-quantification.html#cb227-10"></a>                  <span class="dt">species =</span> <span class="st">&quot;C.elegans&quot;</span>)</span></code></pre></div>
<p>Let’s run the pipeline with two computing nodes.</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="rna-seq-alignment-and-quantification.html#cb228-1"></a>mirRes &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(miRDeep2PL, <span class="dt">outdir =</span> <span class="st">&quot;apps/miRNA/output/&quot;</span>, inputList, paramList,</span>
<span id="cb228-2"><a href="rna-seq-alignment-and-quantification.html#cb228-2"></a>                      <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb228-3"><a href="rna-seq-alignment-and-quantification.html#cb228-3"></a>                                                <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb228-4"><a href="rna-seq-alignment-and-quantification.html#cb228-4"></a>                                                <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">jobname =</span> <span class="st">&quot;miRNA&quot;</span>,</span>
<span id="cb228-5"><a href="rna-seq-alignment-and-quantification.html#cb228-5"></a>                                                                 <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb228-6"><a href="rna-seq-alignment-and-quantification.html#cb228-6"></a>                                                                 <span class="dt">threads =</span> <span class="dv">2</span>),</span>
<span id="cb228-7"><a href="rna-seq-alignment-and-quantification.html#cb228-7"></a>                                                <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">logdir=</span><span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> T),</span>
<span id="cb228-8"><a href="rna-seq-alignment-and-quantification.html#cb228-8"></a>                      <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Here are the results:</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="rna-seq-alignment-and-quantification.html#cb229-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/miRNA/output/sample1&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 10 x 2
##    path                                                                size
##    &lt;fs::path&gt;                                                    &lt;fs::byte&gt;
##  1 apps/miRNA/output/sample1/expression_26_09_2019_t_20_10_36.h…     21.15K
##  2 apps/miRNA/output/sample1/expression_analyses                         4K
##  3 apps/miRNA/output/sample1/miRNAs_expressed_all_samples_26_09…        441
##  4 apps/miRNA/output/sample1/mirna_results_26_09_2019_t_20_10_36         4K
##  5 apps/miRNA/output/sample1/pdfs_26_09_2019_t_20_10_36                  4K
##  6 apps/miRNA/output/sample1/reads_collapsed.arf                     61.36K
##  7 apps/miRNA/output/sample1/reads_collapsed.fa                      59.31K
##  8 apps/miRNA/output/sample1/result_26_09_2019_t_20_10_36.bed           747
##  9 apps/miRNA/output/sample1/result_26_09_2019_t_20_10_36.csv         4.39K
## 10 apps/miRNA/output/sample1/result_26_09_2019_t_20_10_36.html       42.73K</code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dna-seq-variant-calling.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/07-RNASeq.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
