<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Variant calling | Bioinformatics tools and pipelines using R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Variant calling | Bioinformatics tools and pipelines using R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Variant calling | Bioinformatics tools and pipelines using R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and workflows with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu" />
<meta name="author" content="Qian Liu" />


<meta name="date" content="2019-09-26" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="application.html"/>
<link rel="next" href="rna-seq-alignment-and-quantification.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/d3-3.3.8/d3.min.js"></script>
<script src="libs/dagre-0.4.0/dagre-d3.min.js"></script>
<link href="libs/mermaid-0.3.0/dist/mermaid.css" rel="stylesheet" />
<script src="libs/mermaid-0.3.0/dist/mermaid.slim.min.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/chromatography-0.1/chromatography.js"></script>
<script src="libs/DiagrammeR-binding-1.0.1/DiagrammeR.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Rcwl and Case Study</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>0.1</b> R package installation</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>0.2</b> System requirements</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>0.3</b> Docker</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>0.4</b> Structure of the book</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>0.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#common-workflow-language"><i class="fa fa-check"></i><b>1.1</b> Common Workflow Language</a></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#first-example"><i class="fa fa-check"></i><b>1.2</b> First example</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#test-run"><i class="fa fa-check"></i><b>1.3</b> Test run</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="components.html"><a href="components.html"><i class="fa fa-check"></i><b>2</b> Components</a><ul>
<li class="chapter" data-level="2.1" data-path="components.html"><a href="components.html#input-parameters"><i class="fa fa-check"></i><b>2.1</b> Input Parameters</a><ul>
<li class="chapter" data-level="2.1.1" data-path="components.html"><a href="components.html#essential-input-parameters"><i class="fa fa-check"></i><b>2.1.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="2.1.2" data-path="components.html"><a href="components.html#array-inputs"><i class="fa fa-check"></i><b>2.1.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="components.html"><a href="components.html#output-parameters"><i class="fa fa-check"></i><b>2.2</b> Output Parameters</a><ul>
<li class="chapter" data-level="2.2.1" data-path="components.html"><a href="components.html#capturing-output"><i class="fa fa-check"></i><b>2.2.1</b> Capturing Output</a></li>
<li class="chapter" data-level="2.2.2" data-path="components.html"><a href="components.html#array-outputs"><i class="fa fa-check"></i><b>2.2.2</b> Array Outputs</a></li>
<li class="chapter" data-level="2.2.3" data-path="components.html"><a href="components.html#standard-output"><i class="fa fa-check"></i><b>2.2.3</b> Standard output</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>3</b> Writing Pipeline</a><ul>
<li class="chapter" data-level="3.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>3.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="3.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>3.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="run-approaches.html"><a href="run-approaches.html"><i class="fa fa-check"></i><b>4</b> Run approaches</a><ul>
<li class="chapter" data-level="4.1" data-path="run-approaches.html"><a href="run-approaches.html#running-tools-in-docker"><i class="fa fa-check"></i><b>4.1</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="4.2" data-path="run-approaches.html"><a href="run-approaches.html#running-tools-in-cluster-server"><i class="fa fa-check"></i><b>4.2</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="4.3" data-path="run-approaches.html"><a href="run-approaches.html#web-application"><i class="fa fa-check"></i><b>4.3</b> Web Application</a><ul>
<li class="chapter" data-level="4.3.1" data-path="run-approaches.html"><a href="run-approaches.html#cwlparam-example"><i class="fa fa-check"></i><b>4.3.1</b> cwlParam example</a></li>
<li class="chapter" data-level="4.3.2" data-path="run-approaches.html"><a href="run-approaches.html#cwlparam-to-shiny-app"><i class="fa fa-check"></i><b>4.3.2</b> cwlParam to Shiny App</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="application.html"><a href="application.html"><i class="fa fa-check"></i><b>5</b> Application</a><ul>
<li class="chapter" data-level="5.1" data-path="application.html"><a href="application.html#rcwlpipelines-tools"><i class="fa fa-check"></i><b>5.1</b> RcwlPipelines tools</a><ul>
<li class="chapter" data-level="5.1.1" data-path="application.html"><a href="application.html#rcwl-scripts"><i class="fa fa-check"></i><b>5.1.1</b> Rcwl scripts</a></li>
<li class="chapter" data-level="5.1.2" data-path="application.html"><a href="application.html#build-a-pipeline"><i class="fa fa-check"></i><b>5.1.2</b> Build a pipeline</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="application.html"><a href="application.html#rcwlpipelines-summary"><i class="fa fa-check"></i><b>5.2</b> RcwlPipelines summary</a></li>
<li class="chapter" data-level="5.3" data-path="application.html"><a href="application.html#dnaseq-alignment-pipeline"><i class="fa fa-check"></i><b>5.3</b> DNASeq alignment pipeline</a><ul>
<li class="chapter" data-level="5.3.1" data-path="application.html"><a href="application.html#prepare-data"><i class="fa fa-check"></i><b>5.3.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.3.2" data-path="application.html"><a href="application.html#run-in-cluster"><i class="fa fa-check"></i><b>5.3.2</b> Run in cluster</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="application.html"><a href="application.html#rnaseq-pipeline"><i class="fa fa-check"></i><b>5.4</b> RNASeq pipeline</a><ul>
<li class="chapter" data-level="5.4.1" data-path="application.html"><a href="application.html#prepare-data-1"><i class="fa fa-check"></i><b>5.4.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.4.2" data-path="application.html"><a href="application.html#submit-pipeline-with-sge"><i class="fa fa-check"></i><b>5.4.2</b> Submit pipeline with SGE</a></li>
<li class="chapter" data-level="5.4.3" data-path="application.html"><a href="application.html#summarize-qc"><i class="fa fa-check"></i><b>5.4.3</b> Summarize QC</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="application.html"><a href="application.html#mc3-somatic-variant-calling-pipeline"><i class="fa fa-check"></i><b>5.5</b> MC3 somatic variant calling pipeline</a><ul>
<li class="chapter" data-level="5.5.1" data-path="application.html"><a href="application.html#prepare-data-2"><i class="fa fa-check"></i><b>5.5.1</b> Prepare data</a></li>
<li class="chapter" data-level="5.5.2" data-path="application.html"><a href="application.html#run-mc3-pipeline"><i class="fa fa-check"></i><b>5.5.2</b> Run MC3 pipeline</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="application.html"><a href="application.html#gatk4-germline-variant-calling-pipeline"><i class="fa fa-check"></i><b>5.6</b> GATK4 germline variant calling pipeline</a><ul>
<li class="chapter" data-level="5.6.1" data-path="application.html"><a href="application.html#gatk-alignment"><i class="fa fa-check"></i><b>5.6.1</b> GATK Alignment</a></li>
<li class="chapter" data-level="5.6.2" data-path="application.html"><a href="application.html#haplotypecaller"><i class="fa fa-check"></i><b>5.6.2</b> HaplotypeCaller</a></li>
<li class="chapter" data-level="5.6.3" data-path="application.html"><a href="application.html#joint-discovery"><i class="fa fa-check"></i><b>5.6.3</b> Joint Discovery</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="variant-calling.html"><a href="variant-calling.html"><i class="fa fa-check"></i><b>6</b> Variant calling</a><ul>
<li class="chapter" data-level="6.1" data-path="variant-calling.html"><a href="variant-calling.html#dna-alignment"><i class="fa fa-check"></i><b>6.1</b> DNA alignment</a></li>
<li class="chapter" data-level="6.2" data-path="variant-calling.html"><a href="variant-calling.html#germline-variant-calling"><i class="fa fa-check"></i><b>6.2</b> Germline variant calling</a></li>
<li class="chapter" data-level="6.3" data-path="variant-calling.html"><a href="variant-calling.html#somatic-mutation-calling"><i class="fa fa-check"></i><b>6.3</b> Somatic mutation calling</a><ul>
<li class="chapter" data-level="6.3.1" data-path="variant-calling.html"><a href="variant-calling.html#mutect2"><i class="fa fa-check"></i><b>6.3.1</b> MuTect2</a></li>
<li class="chapter" data-level="6.3.2" data-path="variant-calling.html"><a href="variant-calling.html#muse"><i class="fa fa-check"></i><b>6.3.2</b> MuSE</a></li>
<li class="chapter" data-level="6.3.3" data-path="variant-calling.html"><a href="variant-calling.html#strelka2"><i class="fa fa-check"></i><b>6.3.3</b> Strelka2</a></li>
<li class="chapter" data-level="6.3.4" data-path="variant-calling.html"><a href="variant-calling.html#somaticsniper"><i class="fa fa-check"></i><b>6.3.4</b> SomaticSniper</a></li>
<li class="chapter" data-level="6.3.5" data-path="variant-calling.html"><a href="variant-calling.html#vardict"><i class="fa fa-check"></i><b>6.3.5</b> VarDict</a></li>
<li class="chapter" data-level="6.3.6" data-path="variant-calling.html"><a href="variant-calling.html#lofreq"><i class="fa fa-check"></i><b>6.3.6</b> LoFreq</a></li>
<li class="chapter" data-level="6.3.7" data-path="variant-calling.html"><a href="variant-calling.html#varscan2"><i class="fa fa-check"></i><b>6.3.7</b> VarScan2</a></li>
<li class="chapter" data-level="6.3.8" data-path="variant-calling.html"><a href="variant-calling.html#lancet"><i class="fa fa-check"></i><b>6.3.8</b> Lancet</a></li>
<li class="chapter" data-level="6.3.9" data-path="variant-calling.html"><a href="variant-calling.html#neusomatic"><i class="fa fa-check"></i><b>6.3.9</b> Neusomatic</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html"><i class="fa fa-check"></i><b>7</b> RNA-Seq alignment and quantification</a><ul>
<li class="chapter" data-level="7.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#bulk-mrna"><i class="fa fa-check"></i><b>7.1</b> bulk mRNA</a></li>
<li class="chapter" data-level="7.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#transcriptome-quantification"><i class="fa fa-check"></i><b>7.2</b> transcriptome quantification</a></li>
<li class="chapter" data-level="7.3" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#single-cell-rnaseq"><i class="fa fa-check"></i><b>7.3</b> single cell RNAseq</a></li>
<li class="chapter" data-level="7.4" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#mirna"><i class="fa fa-check"></i><b>7.4</b> miRNA</a><ul>
<li class="chapter" data-level="7.4.1" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#prepare-reference"><i class="fa fa-check"></i><b>7.4.1</b> Prepare reference</a></li>
<li class="chapter" data-level="7.4.2" data-path="rna-seq-alignment-and-quantification.html"><a href="rna-seq-alignment-and-quantification.html#run-mirdeep2-pipeline"><i class="fa fa-check"></i><b>7.4.2</b> Run miRDeep2 pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines using R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="variant-calling" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Variant calling</h1>
<p>First, we load the required packages.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="variant-calling.html#cb151-1"></a><span class="kw">library</span>(RcwlPipelines)</span>
<span id="cb151-2"><a href="variant-calling.html#cb151-2"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb151-3"><a href="variant-calling.html#cb151-3"></a><span class="kw">library</span>(fs)</span>
<span id="cb151-4"><a href="variant-calling.html#cb151-4"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb151-5"><a href="variant-calling.html#cb151-5"></a><span class="kw">library</span>(jsonlite)</span></code></pre></div>
<div id="dna-alignment" class="section level2">
<h2><span class="header-section-number">6.1</span> DNA alignment</h2>
<p>We use the test data sets from the ICGC-TCGA DREAM Mutation Calling challenge for demonstration and performance test. The test sets contains two pairs of BAM files with SNVs and SVs in chr19 and chr20 separately. The ‘Truth’ VCF files are also included.</p>
<p>Here is the URL to download the test data.</p>
<ul>
<li><a href="https://www.synapse.org/#!Synapse:syn2335184" class="uri">https://www.synapse.org/#!Synapse:syn2335184</a></li>
</ul>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="variant-calling.html#cb152-1"></a><span class="kw">list.files</span>(<span class="st">&quot;apps/data/DREAM/&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;chr19.normal.bam&quot;       &quot;chr19.normal.bam.bai&quot;  
## [3] &quot;chr19.truth.vcf.gz&quot;     &quot;chr19.truth.vcf.gz.tbi&quot;
## [5] &quot;chr19.tumor.bam&quot;        &quot;chr19.tumor.bam.bai&quot;   
## [7] &quot;normal.chr20.bam&quot;       &quot;truth.chr20.vcf.gz&quot;    
## [9] &quot;tumor.chr20.bam&quot;</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="variant-calling.html#cb154-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/data/DREAM&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 9 x 2
##   path                                          size
##   &lt;fs::path&gt;                             &lt;fs::bytes&gt;
## 1 apps/data/DREAM/chr19.normal.bam             3.31G
## 2 apps/data/DREAM/chr19.normal.bam.bai       171.64K
## 3 apps/data/DREAM/chr19.truth.vcf.gz           21.6K
## 4 apps/data/DREAM/chr19.truth.vcf.gz.tbi      11.27K
## 5 apps/data/DREAM/chr19.tumor.bam              3.38G
## 6 apps/data/DREAM/chr19.tumor.bam.bai          1.45M
## 7 apps/data/DREAM/normal.chr20.bam             2.93G
## 8 apps/data/DREAM/truth.chr20.vcf.gz          16.58K
## 9 apps/data/DREAM/tumor.chr20.bam               2.9G</code></pre>
<p>In order to show the variant calling workflow from the very beginning. We convert the BAM files back to fastqs using <code>picard SamToFastq</code> and treat the two BAM files as samples from two patients.</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="variant-calling.html#cb156-1"></a><span class="kw">inputs</span>(SamToFastq)</span></code></pre></div>
<pre><code>## inputs:
##   bam (File): I= 
##   fq1 (string): F= 
##   fq2 (string): F2=</code></pre>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="variant-calling.html#cb158-1"></a><span class="co">## prepare inputs</span></span>
<span id="cb158-2"><a href="variant-calling.html#cb158-2"></a>bams &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&quot;apps/data/DREAM/&quot;</span>, <span class="st">&quot;.bam$&quot;</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb158-3"><a href="variant-calling.html#cb158-3"></a>samples &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;normal19&quot;</span>, <span class="st">&quot;tumor19&quot;</span>, <span class="st">&quot;normal20&quot;</span>, <span class="st">&quot;tumor20&quot;</span>)</span>
<span id="cb158-4"><a href="variant-calling.html#cb158-4"></a>Bams &lt;-<span class="st"> </span><span class="kw">tapply</span>(bams, samples, as.list)</span>
<span id="cb158-5"><a href="variant-calling.html#cb158-5"></a>fq1 &lt;-<span class="st"> </span><span class="kw">tapply</span>(<span class="kw">paste0</span>(samples, <span class="st">&quot;_R1.fq&quot;</span>), samples, as.list)</span>
<span id="cb158-6"><a href="variant-calling.html#cb158-6"></a>fq2 &lt;-<span class="st"> </span><span class="kw">tapply</span>(<span class="kw">paste0</span>(samples, <span class="st">&quot;_R2.fq&quot;</span>), samples, as.list)</span>
<span id="cb158-7"><a href="variant-calling.html#cb158-7"></a></span>
<span id="cb158-8"><a href="variant-calling.html#cb158-8"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">bam =</span> Bams,</span>
<span id="cb158-9"><a href="variant-calling.html#cb158-9"></a>                  <span class="dt">fq1 =</span> fq1,</span>
<span id="cb158-10"><a href="variant-calling.html#cb158-10"></a>                  <span class="dt">fq2 =</span> fq2)</span>
<span id="cb158-11"><a href="variant-calling.html#cb158-11"></a><span class="co">## run in HPC using SGE</span></span>
<span id="cb158-12"><a href="variant-calling.html#cb158-12"></a>res1 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(SamToFastq, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/fastq/&quot;</span>, inputList,</span>
<span id="cb158-13"><a href="variant-calling.html#cb158-13"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="kw">length</span>(samples), <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb158-14"><a href="variant-calling.html#cb158-14"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb158-15"><a href="variant-calling.html#cb158-15"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">jobname =</span> <span class="st">&quot;sam2fq&quot;</span>,</span>
<span id="cb158-16"><a href="variant-calling.html#cb158-16"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb158-17"><a href="variant-calling.html#cb158-17"></a>                                                               <span class="dt">threads =</span> <span class="dv">16</span>),</span>
<span id="cb158-18"><a href="variant-calling.html#cb158-18"></a>                                              <span class="dt">log=</span><span class="ot">TRUE</span>, <span class="dt">logdir=</span><span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> T),</span>
<span id="cb158-19"><a href="variant-calling.html#cb158-19"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Let’s check the outputs.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="variant-calling.html#cb159-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/fastq/&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>, <span class="dt">glob =</span> <span class="st">&quot;*.fq&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 8 x 2
##   path                                               size
##   &lt;fs::path&gt;                                  &lt;fs::bytes&gt;
## 1 apps/variants/fastq/normal19/normal19_R1.fq       2.17G
## 2 apps/variants/fastq/normal19/normal19_R2.fq       2.17G
## 3 apps/variants/fastq/normal20/normal20_R1.fq       2.14G
## 4 apps/variants/fastq/normal20/normal20_R2.fq       2.14G
## 5 apps/variants/fastq/tumor19/tumor19_R1.fq         2.18G
## 6 apps/variants/fastq/tumor19/tumor19_R2.fq         2.18G
## 7 apps/variants/fastq/tumor20/tumor20_R1.fq         2.14G
## 8 apps/variants/fastq/tumor20/tumor20_R2.fq         2.14G</code></pre>
<p>Now, we are ready to run the DNA alignment pipeline, <code>bwaMRecal</code>, including all necessary steps to prepare the analysis-ready BAM files. We can check the required inputs and steps.</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="variant-calling.html#cb161-1"></a><span class="kw">inputs</span>(bwaMRecal)</span></code></pre></div>
<pre><code>## inputs:
##   outBam (string):  
##   RG (string):  
##   threads (int):  
##   Ref (File):  
##   FQ1s (File):  
##   FQ2s (File):  
##   knowSites:
##     type: array
##     prefix:</code></pre>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="variant-calling.html#cb163-1"></a><span class="kw">names</span>(<span class="kw">steps</span>(bwaMRecal))</span></code></pre></div>
<pre><code>## [1] &quot;bwaAlign&quot;  &quot;markdup&quot;   &quot;BaseRecal&quot;</code></pre>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="variant-calling.html#cb165-1"></a>fqs &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&quot;apps/variants/fastq&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb165-2"><a href="variant-calling.html#cb165-2"></a>fq1 &lt;-<span class="st"> </span>fqs[<span class="kw">grep</span>(<span class="st">&quot;R1&quot;</span>, fqs)]</span>
<span id="cb165-3"><a href="variant-calling.html#cb165-3"></a>fq2 &lt;-<span class="st"> </span>fqs[<span class="kw">grep</span>(<span class="st">&quot;R2&quot;</span>, fqs)]</span>
<span id="cb165-4"><a href="variant-calling.html#cb165-4"></a>ids &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;_.*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">basename</span>(fq1))</span>
<span id="cb165-5"><a href="variant-calling.html#cb165-5"></a>fq1L &lt;-<span class="st"> </span><span class="kw">tapply</span>(fq1, ids, as.list)</span>
<span id="cb165-6"><a href="variant-calling.html#cb165-6"></a>fq2L &lt;-<span class="st"> </span><span class="kw">tapply</span>(fq2, ids, as.list)</span>
<span id="cb165-7"><a href="variant-calling.html#cb165-7"></a></span>
<span id="cb165-8"><a href="variant-calling.html#cb165-8"></a>RGs &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;@RG&quot;</span>,</span>
<span id="cb165-9"><a href="variant-calling.html#cb165-9"></a>             <span class="kw">paste0</span>(<span class="st">&quot;ID:&quot;</span>, ids),</span>
<span id="cb165-10"><a href="variant-calling.html#cb165-10"></a>             <span class="kw">paste0</span>(<span class="st">&quot;LB:&quot;</span>, ids),</span>
<span id="cb165-11"><a href="variant-calling.html#cb165-11"></a>             <span class="kw">paste0</span>(<span class="st">&quot;DT:&quot;</span>, <span class="kw">Sys.Date</span>()),</span>
<span id="cb165-12"><a href="variant-calling.html#cb165-12"></a>             <span class="kw">paste0</span>(<span class="st">&quot;PL:&quot;</span>, <span class="st">&quot;Illumina&quot;</span>),</span>
<span id="cb165-13"><a href="variant-calling.html#cb165-13"></a>             <span class="st">&quot;CN:RCWL&quot;</span>,</span>
<span id="cb165-14"><a href="variant-calling.html#cb165-14"></a>             <span class="kw">paste0</span>(<span class="st">&quot;SM:&quot;</span>, ids), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\\</span><span class="st">t&quot;</span>)</span>
<span id="cb165-15"><a href="variant-calling.html#cb165-15"></a>RGL &lt;-<span class="st"> </span><span class="kw">tapply</span>(RGs, ids, as.list)</span>
<span id="cb165-16"><a href="variant-calling.html#cb165-16"></a>outBam &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">paste0</span>(<span class="kw">names</span>(RGL), <span class="st">&quot;.bam&quot;</span>))</span>
<span id="cb165-17"><a href="variant-calling.html#cb165-17"></a></span>
<span id="cb165-18"><a href="variant-calling.html#cb165-18"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">RG =</span> RGL,</span>
<span id="cb165-19"><a href="variant-calling.html#cb165-19"></a>                  <span class="dt">outBam =</span> outBam,</span>
<span id="cb165-20"><a href="variant-calling.html#cb165-20"></a>                  <span class="dt">FQ1s =</span> fq1L,</span>
<span id="cb165-21"><a href="variant-calling.html#cb165-21"></a>                  <span class="dt">FQ2s =</span> fq2L)</span>
<span id="cb165-22"><a href="variant-calling.html#cb165-22"></a></span>
<span id="cb165-23"><a href="variant-calling.html#cb165-23"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb165-24"><a href="variant-calling.html#cb165-24"></a>                  <span class="dt">Ref =</span><span class="st">&quot;apps/data/hs37d5.fa&quot;</span>,</span>
<span id="cb165-25"><a href="variant-calling.html#cb165-25"></a>                  <span class="dt">knowSites =</span> <span class="kw">list</span>(</span>
<span id="cb165-26"><a href="variant-calling.html#cb165-26"></a>                      <span class="st">&quot;apps/data/bundle/dbsnp_138.b37.vcf&quot;</span>,</span>
<span id="cb165-27"><a href="variant-calling.html#cb165-27"></a>                      <span class="st">&quot;apps/data/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf&quot;</span></span>
<span id="cb165-28"><a href="variant-calling.html#cb165-28"></a>                  ))</span>
<span id="cb165-29"><a href="variant-calling.html#cb165-29"></a></span>
<span id="cb165-30"><a href="variant-calling.html#cb165-30"></a>res2 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(bwaMRecal, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/BAM&quot;</span>, inputList, paramList,</span>
<span id="cb165-31"><a href="variant-calling.html#cb165-31"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="kw">length</span>(samples),</span>
<span id="cb165-32"><a href="variant-calling.html#cb165-32"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb165-33"><a href="variant-calling.html#cb165-33"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb165-34"><a href="variant-calling.html#cb165-34"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb165-35"><a href="variant-calling.html#cb165-35"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb165-36"><a href="variant-calling.html#cb165-36"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;align&quot;</span>),</span>
<span id="cb165-37"><a href="variant-calling.html#cb165-37"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb165-38"><a href="variant-calling.html#cb165-38"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Here are the outputs:</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="variant-calling.html#cb166-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/BAM/&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 26 x 2
##    path                                                       size
##    &lt;fs::path&gt;                                          &lt;fs::bytes&gt;
##  1 apps/variants/BAM/normal19                                   4K
##  2 apps/variants/BAM/normal19/normal19.bai                    2.6M
##  3 apps/variants/BAM/normal19/normal19.bam                    2.3G
##  4 apps/variants/BAM/normal19/normal19.bam.bai                2.6M
##  5 apps/variants/BAM/normal19/normal19.bam.markdup.txt       2.75K
##  6 apps/variants/BAM/normal19/normal19.flagstat.txt            448
##  7 apps/variants/BAM/normal19/normal19.stats.txt            95.16K
##  8 apps/variants/BAM/normal20                                   4K
##  9 apps/variants/BAM/normal20/normal20.bai                   2.45M
## 10 apps/variants/BAM/normal20/normal20.bam                   2.23G
## # … with 16 more rows</code></pre>
</div>
<div id="germline-variant-calling" class="section level2">
<h2><span class="header-section-number">6.2</span> Germline variant calling</h2>
<p>The GATK germline variant calling pipeline is one of the most popular method to call germline short variants. The latest best practice was implemented with the Workflow Description Language (WDL). The WDL scripts was simply wrapped with <code>Rcwl</code> in the <code>RcwlPipelines</code> package.</p>
<ul>
<li><a href="https://github.com/gatk-workflows/gatk4-germline-snps-indels" class="uri">https://github.com/gatk-workflows/gatk4-germline-snps-indels</a></li>
</ul>
<p>There are two steps to run the latest GATK pipeline if the BAMs are prepared. First the <code>haplotypecaller</code> step is used to call variants by samples. The “WDL” script and “json” input for this step are also contained in the package.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="variant-calling.html#cb168-1"></a>wdl1 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.wdl&quot;</span>)</span>
<span id="cb168-2"><a href="variant-calling.html#cb168-2"></a>tmpl1 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/haplotypecaller-gvcf-gatk4.hg38.inputs.local.json&quot;</span>)</span></code></pre></div>
<p>We need to modify the reference from “hg38” to “b37”. Please note that all the file paths should be absolute paths for the “WDL” inputs.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="variant-calling.html#cb169-1"></a>json1 &lt;-<span class="st"> </span><span class="kw">fromJSON</span>(tmpl1)</span>
<span id="cb169-2"><a href="variant-calling.html#cb169-2"></a>json1<span class="op">$</span>HaplotypeCallerGvcf_GATK4.ref_dict &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/hs37d5.dict&quot;</span>)</span>
<span id="cb169-3"><a href="variant-calling.html#cb169-3"></a>json1<span class="op">$</span>HaplotypeCallerGvcf_GATK4.ref_fasta &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/hs37d5.fa&quot;</span>)</span>
<span id="cb169-4"><a href="variant-calling.html#cb169-4"></a>json1<span class="op">$</span>HaplotypeCallerGvcf_GATK4.ref_fasta_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/hs37d5.fa.fai&quot;</span>)</span>
<span id="cb169-5"><a href="variant-calling.html#cb169-5"></a><span class="kw">dir.create</span>(<span class="st">&quot;apps/variants/vcf&quot;</span>)</span>
<span id="cb169-6"><a href="variant-calling.html#cb169-6"></a><span class="kw">writeLines</span>(jsonlite<span class="op">::</span><span class="kw">toJSON</span>(json1, <span class="dt">pretty =</span> <span class="ot">TRUE</span>, <span class="dt">auto_unbox =</span> T), <span class="st">&quot;apps/variants/vcf/tmpl1.json&quot;</span>)</span></code></pre></div>
<p>A interval list is required to define the exome capture region. We couldn’t find one for the TCGA data, so here we just simply use the whole chromosome regions.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="variant-calling.html#cb170-1"></a>fai &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;apps/data/hs37d5.fa.fai&quot;</span>)</span>
<span id="cb170-2"><a href="variant-calling.html#cb170-2"></a>fai &lt;-<span class="st"> </span>fai[fai[,<span class="dv">1</span>] <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">19</span>, <span class="dv">20</span>),]</span>
<span id="cb170-3"><a href="variant-calling.html#cb170-3"></a>bed &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">as.character</span>(fai[,<span class="dv">1</span>]), <span class="dv">1</span>, fai[,<span class="dv">2</span>])</span>
<span id="cb170-4"><a href="variant-calling.html#cb170-4"></a><span class="kw">write.table</span>(bed, <span class="st">&quot;apps/variants/vcf/intval.bed&quot;</span>,</span>
<span id="cb170-5"><a href="variant-calling.html#cb170-5"></a>            <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">col.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)</span>
<span id="cb170-6"><a href="variant-calling.html#cb170-6"></a></span>
<span id="cb170-7"><a href="variant-calling.html#cb170-7"></a>BedToIntervalList<span class="op">$</span>bed &lt;-<span class="st"> &quot;apps/variants/vcf/intval.bed&quot;</span></span>
<span id="cb170-8"><a href="variant-calling.html#cb170-8"></a>BedToIntervalList<span class="op">$</span>SD &lt;-<span class="st"> &quot;apps/data/hs37d5.dict&quot;</span></span>
<span id="cb170-9"><a href="variant-calling.html#cb170-9"></a>BedToIntervalList<span class="op">$</span>out &lt;-<span class="st"> &quot;region.interval_list&quot;</span></span>
<span id="cb170-10"><a href="variant-calling.html#cb170-10"></a>r1 &lt;-<span class="st"> </span><span class="kw">runCWL</span>(BedToIntervalList, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/vcf/&quot;</span>)</span>
<span id="cb170-11"><a href="variant-calling.html#cb170-11"></a></span>
<span id="cb170-12"><a href="variant-calling.html#cb170-12"></a><span class="kw">writeLines</span>(<span class="kw">normalizePath</span>(<span class="st">&quot;apps/variants/vcf/region.interval_list&quot;</span>), <span class="st">&quot;apps/variants/vcf/intval.txt&quot;</span>)</span>
<span id="cb170-13"><a href="variant-calling.html#cb170-13"></a><span class="kw">writeLines</span>(<span class="kw">paste0</span>(bed[,<span class="dv">1</span>], <span class="st">&quot;:&quot;</span>, bed[,<span class="dv">2</span>], <span class="st">&quot;-&quot;</span>, bed[,<span class="dv">3</span>]), <span class="st">&quot;apps/variants/vcf/intval.unpad.txt&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="variant-calling.html#cb171-1"></a><span class="co">## fix bai paths</span></span>
<span id="cb171-2"><a href="variant-calling.html#cb171-2"></a><span class="kw">file.copy</span>(<span class="st">&quot;apps/variants/BAM/normal19/normal19.bam.bai&quot;</span>,</span>
<span id="cb171-3"><a href="variant-calling.html#cb171-3"></a>          <span class="st">&quot;apps/variants/BAM/normal19/normal19.bai&quot;</span>)</span>
<span id="cb171-4"><a href="variant-calling.html#cb171-4"></a><span class="kw">file.copy</span>(<span class="st">&quot;apps/variants/BAM/normal20/normal20.bam.bai&quot;</span>,</span>
<span id="cb171-5"><a href="variant-calling.html#cb171-5"></a>          <span class="st">&quot;apps/variants/BAM/normal20/normal20.bai&quot;</span>)</span>
<span id="cb171-6"><a href="variant-calling.html#cb171-6"></a></span>
<span id="cb171-7"><a href="variant-calling.html#cb171-7"></a>bams &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/BAM&quot;</span>, <span class="st">&quot;normal.*.bam$&quot;</span>,</span>
<span id="cb171-8"><a href="variant-calling.html#cb171-8"></a>                                 <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb171-9"><a href="variant-calling.html#cb171-9"></a>bams &lt;-<span class="st"> </span><span class="kw">as.list</span>(bams)</span>
<span id="cb171-10"><a href="variant-calling.html#cb171-10"></a><span class="kw">names</span>(bams) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;normal19&quot;</span>, <span class="st">&quot;normal20&quot;</span>)</span>
<span id="cb171-11"><a href="variant-calling.html#cb171-11"></a></span>
<span id="cb171-12"><a href="variant-calling.html#cb171-12"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">bam =</span> bams)</span>
<span id="cb171-13"><a href="variant-calling.html#cb171-13"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">intervals =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/variants/vcf/intval.txt&quot;</span>),</span>
<span id="cb171-14"><a href="variant-calling.html#cb171-14"></a>                  <span class="dt">cromwell =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/cromwell-45.jar&quot;</span>),</span>
<span id="cb171-15"><a href="variant-calling.html#cb171-15"></a>                  <span class="dt">wdl =</span> wdl1,</span>
<span id="cb171-16"><a href="variant-calling.html#cb171-16"></a>                  <span class="dt">tmpl =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/variants/vcf/tmpl1.json&quot;</span>))</span>
<span id="cb171-17"><a href="variant-calling.html#cb171-17"></a></span>
<span id="cb171-18"><a href="variant-calling.html#cb171-18"></a>res3 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(hapCall, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/vcf&quot;</span>, inputList, paramList,</span>
<span id="cb171-19"><a href="variant-calling.html#cb171-19"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="kw">length</span>(bams),</span>
<span id="cb171-20"><a href="variant-calling.html#cb171-20"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb171-21"><a href="variant-calling.html#cb171-21"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb171-22"><a href="variant-calling.html#cb171-22"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb171-23"><a href="variant-calling.html#cb171-23"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb171-24"><a href="variant-calling.html#cb171-24"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;hapCall&quot;</span>),</span>
<span id="cb171-25"><a href="variant-calling.html#cb171-25"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb171-26"><a href="variant-calling.html#cb171-26"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>, <span class="dt">cwlTemp=</span>T)</span></code></pre></div>
<p>Here are the results.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="variant-calling.html#cb172-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/vcf/&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>, <span class="dt">glob =</span> <span class="st">&quot;*.g.vcf.gz&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 2 x 2
##   path                                                       size
##   &lt;fs::path&gt;                                          &lt;fs::bytes&gt;
## 1 apps/variants/vcf/normal19/output/normal19.g.vcf.gz        192M
## 2 apps/variants/vcf/normal20/output/normal20.g.vcf.gz        192M</code></pre>
<p>Then we perform joint genotyping with <code>jdCall</code>. The reference genome bundles are also need to be changed to “b37”. You can skip this step if you already have a “b37” version of metadata file.</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="variant-calling.html#cb174-1"></a>wdl2 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.wdl&quot;</span>)</span>
<span id="cb174-2"><a href="variant-calling.html#cb174-2"></a>tmpl2 &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;RcwlPipelines&quot;</span>, <span class="st">&quot;GATK4/gatk4-germline-snps-indels/joint-discovery-gatk4-local.hg38.wgs.inputs.json&quot;</span>)</span>
<span id="cb174-3"><a href="variant-calling.html#cb174-3"></a><span class="co">## change to b37</span></span>
<span id="cb174-4"><a href="variant-calling.html#cb174-4"></a>json2 &lt;-<span class="st"> </span><span class="kw">fromJSON</span>(tmpl2)</span>
<span id="cb174-5"><a href="variant-calling.html#cb174-5"></a>json2<span class="op">$</span>JointGenotyping.ref_dict &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/hs37d5.dict&quot;</span>)</span>
<span id="cb174-6"><a href="variant-calling.html#cb174-6"></a>json2<span class="op">$</span>JointGenotyping.ref_fasta &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/hs37d5.fa&quot;</span>)</span>
<span id="cb174-7"><a href="variant-calling.html#cb174-7"></a>json2<span class="op">$</span>JointGenotyping.ref_fasta_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/hs37d5.fa.fai&quot;</span>)</span>
<span id="cb174-8"><a href="variant-calling.html#cb174-8"></a>json2<span class="op">$</span>JointGenotyping.dbsnp_vcf &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/dbsnp_138.b37.vcf&quot;</span>)</span>
<span id="cb174-9"><a href="variant-calling.html#cb174-9"></a>json2<span class="op">$</span>JointGenotyping.dbsnp_vcf_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/dbsnp_138.b37.vcf.idx&quot;</span>)</span>
<span id="cb174-10"><a href="variant-calling.html#cb174-10"></a>json2<span class="op">$</span>JointGenotyping.one_thousand_genomes_resource_vcf &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/1000G_phase1.snps.high_confidence.b37.vcf&quot;</span>)</span>
<span id="cb174-11"><a href="variant-calling.html#cb174-11"></a>json2<span class="op">$</span>JointGenotyping.one_thousand_genomes_resource_vcf_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/1000G_phase1.snps.high_confidence.b37.vcf.idx&quot;</span>)</span>
<span id="cb174-12"><a href="variant-calling.html#cb174-12"></a>json2<span class="op">$</span>JointGenotyping.omni_resource_vcf &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/1000G_omni2.5.b37.vcf&quot;</span>)</span>
<span id="cb174-13"><a href="variant-calling.html#cb174-13"></a>json2<span class="op">$</span>JointGenotyping.omni_resource_vcf_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/1000G_omni2.5.b37.vcf.idx&quot;</span>)</span>
<span id="cb174-14"><a href="variant-calling.html#cb174-14"></a>json2<span class="op">$</span>JointGenotyping.mills_resource_vcf &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf&quot;</span>)</span>
<span id="cb174-15"><a href="variant-calling.html#cb174-15"></a>json2<span class="op">$</span>JointGenotyping.mills_resource_vcf_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf.idx&quot;</span>)</span>
<span id="cb174-16"><a href="variant-calling.html#cb174-16"></a>json2<span class="op">$</span>JointGenotyping.axiomPoly_resource_vcf &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/Axiom_Exome_Plus.genotypes.all_populations.poly.vcf.gz&quot;</span>)</span>
<span id="cb174-17"><a href="variant-calling.html#cb174-17"></a>json2<span class="op">$</span>JointGenotyping.axiomPoly_resource_vcf_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/Axiom_Exome_Plus.genotypes.all_populations.poly.vcf.gz.tbi&quot;</span>)</span>
<span id="cb174-18"><a href="variant-calling.html#cb174-18"></a>json2<span class="op">$</span>JointGenotyping.hapmap_resource_vcf &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/hapmap_3.3.b37.vcf&quot;</span>)</span>
<span id="cb174-19"><a href="variant-calling.html#cb174-19"></a>json2<span class="op">$</span>JointGenotyping.hapmap_resource_vcf_index &lt;-<span class="st"> </span><span class="kw">normalizePath</span>(<span class="st">&quot;apps/data/bundle/hapmap_3.3.b37.vcf.idx&quot;</span>)</span>
<span id="cb174-20"><a href="variant-calling.html#cb174-20"></a></span>
<span id="cb174-21"><a href="variant-calling.html#cb174-21"></a><span class="kw">writeLines</span>(<span class="kw">toJSON</span>(json2, <span class="dt">pretty =</span> <span class="ot">TRUE</span>, <span class="dt">auto_unbox =</span> T), <span class="st">&quot;apps/variants/vcf/tmpl2.json&quot;</span>)</span></code></pre></div>
<p>The joint calling step takes the inputs from the previous step and run together.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="variant-calling.html#cb175-1"></a><span class="co">## prepare inputs</span></span>
<span id="cb175-2"><a href="variant-calling.html#cb175-2"></a>sampleName &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">names</span>(bams), <span class="dt">collapse =</span> <span class="st">&quot;,&quot;</span>)</span>
<span id="cb175-3"><a href="variant-calling.html#cb175-3"></a>gvcf &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">normalizePath</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/vcf&quot;</span>, <span class="st">&quot;.gz$&quot;</span>,</span>
<span id="cb175-4"><a href="variant-calling.html#cb175-4"></a>                                       <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)),</span>
<span id="cb175-5"><a href="variant-calling.html#cb175-5"></a>              <span class="dt">collapse =</span> <span class="st">&quot;,&quot;</span>)</span>
<span id="cb175-6"><a href="variant-calling.html#cb175-6"></a></span>
<span id="cb175-7"><a href="variant-calling.html#cb175-7"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sampleName =</span> <span class="kw">list</span>(<span class="dt">res =</span> sampleName),</span>
<span id="cb175-8"><a href="variant-calling.html#cb175-8"></a>                  <span class="dt">gvcf =</span> <span class="kw">list</span>(<span class="dt">res =</span> gvcf))</span>
<span id="cb175-9"><a href="variant-calling.html#cb175-9"></a></span>
<span id="cb175-10"><a href="variant-calling.html#cb175-10"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">callsetName =</span> <span class="st">&quot;germline&quot;</span>,</span>
<span id="cb175-11"><a href="variant-calling.html#cb175-11"></a>                  <span class="dt">intervals =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/variants/vcf/region.interval_list&quot;</span>),</span>
<span id="cb175-12"><a href="variant-calling.html#cb175-12"></a>                  <span class="dt">unpadded_intervals =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/variants/vcf/intval.unpad.txt&quot;</span>),</span>
<span id="cb175-13"><a href="variant-calling.html#cb175-13"></a>                  <span class="dt">tmpl =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/variants/vcf/tmpl2.json&quot;</span>),</span>
<span id="cb175-14"><a href="variant-calling.html#cb175-14"></a>                  <span class="dt">wdl =</span> wdl2,</span>
<span id="cb175-15"><a href="variant-calling.html#cb175-15"></a>                  <span class="dt">cromwell =</span> <span class="kw">normalizePath</span>(<span class="st">&quot;apps/cromwell-45.jar&quot;</span>))</span>
<span id="cb175-16"><a href="variant-calling.html#cb175-16"></a>res4 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(jdCall, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/vcf&quot;</span>, inputList, paramList,</span>
<span id="cb175-17"><a href="variant-calling.html#cb175-17"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">1</span>,</span>
<span id="cb175-18"><a href="variant-calling.html#cb175-18"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb175-19"><a href="variant-calling.html#cb175-19"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb175-20"><a href="variant-calling.html#cb175-20"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb175-21"><a href="variant-calling.html#cb175-21"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb175-22"><a href="variant-calling.html#cb175-22"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;jdCall&quot;</span>),</span>
<span id="cb175-23"><a href="variant-calling.html#cb175-23"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb175-24"><a href="variant-calling.html#cb175-24"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>, <span class="dt">cwlTemp=</span>T)</span></code></pre></div>
<p>Here are the final results:</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="variant-calling.html#cb176-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/vcf/res/output&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 5 x 2
##   path                                                                 size
##   &lt;fs::path&gt;                                                     &lt;fs::byte&gt;
## 1 apps/variants/vcf/res/output/germline.variant_calling_detail_…       1.9K
## 2 apps/variants/vcf/res/output/germline.variant_calling_summary…      1.63K
## 3 apps/variants/vcf/res/output/germline.vcf.gz                         9.4M
## 4 apps/variants/vcf/res/output/germline.vcf.gz.tbi                   37.64K
## 5 apps/variants/vcf/res/output/out.intervals                             27</code></pre>
</div>
<div id="somatic-mutation-calling" class="section level2">
<h2><span class="header-section-number">6.3</span> Somatic mutation calling</h2>
<p>We have collected many somatic mutation callers, including MuTect2, strelka2, MuSe, VarDict, SomaticSniper, LoFreq, VarScan2, lancet and neusomatic. Recent research papers have reported that the combination of multiple callers could improve the sensitivity and specificity.</p>
<div id="mutect2" class="section level3">
<h3><span class="header-section-number">6.3.1</span> MuTect2</h3>
<p>The MuTect2 was built based on the latest best practise from GATK 4.
<a href="https://software.broadinstitute.org/gatk/best-practices/workflow?id=11146" class="uri">https://software.broadinstitute.org/gatk/best-practices/workflow?id=11146</a></p>
<p>The required resources files can be downloaded from its google storage. &lt;gs://gatk-best-practices&gt;</p>
<ol style="list-style-type: decimal">
<li>Call variants on normal samples</li>
</ol>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="variant-calling.html#cb178-1"></a>bams &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&quot;apps/variants/BAM&quot;</span>, <span class="st">&quot;.bam$&quot;</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb178-2"><a href="variant-calling.html#cb178-2"></a>nbam &lt;-<span class="st"> </span><span class="kw">as.list</span>(bams[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</span>
<span id="cb178-3"><a href="variant-calling.html#cb178-3"></a><span class="kw">names</span>(nbam) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;normal19&quot;</span>, <span class="st">&quot;normal20&quot;</span>)</span>
<span id="cb178-4"><a href="variant-calling.html#cb178-4"></a>ovcf &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">paste0</span>(<span class="kw">names</span>(nbam), <span class="st">&quot;.vcf&quot;</span>))</span>
<span id="cb178-5"><a href="variant-calling.html#cb178-5"></a></span>
<span id="cb178-6"><a href="variant-calling.html#cb178-6"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">tbam =</span> nbam,</span>
<span id="cb178-7"><a href="variant-calling.html#cb178-7"></a>                  <span class="dt">out =</span> ovcf)</span>
<span id="cb178-8"><a href="variant-calling.html#cb178-8"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">Ref =</span> <span class="st">&quot;apps/data/hs37d5.fa&quot;</span>,</span>
<span id="cb178-9"><a href="variant-calling.html#cb178-9"></a>                  <span class="dt">interval =</span> <span class="st">&quot;apps/variants/vcf/intval.bed&quot;</span>)</span>
<span id="cb178-10"><a href="variant-calling.html#cb178-10"></a><span class="co">## avoid a bug in GenomicsDBImport</span></span>
<span id="cb178-11"><a href="variant-calling.html#cb178-11"></a><span class="kw">arguments</span>(Mutect2) &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;-max-mnp-distance&quot;</span>, <span class="st">&quot;0&quot;</span>)</span>
<span id="cb178-12"><a href="variant-calling.html#cb178-12"></a>res5 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(Mutect2, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/mutect2&quot;</span>, inputList, paramList,</span>
<span id="cb178-13"><a href="variant-calling.html#cb178-13"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb178-14"><a href="variant-calling.html#cb178-14"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb178-15"><a href="variant-calling.html#cb178-15"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb178-16"><a href="variant-calling.html#cb178-16"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb178-17"><a href="variant-calling.html#cb178-17"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb178-18"><a href="variant-calling.html#cb178-18"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;mutect2&quot;</span>),</span>
<span id="cb178-19"><a href="variant-calling.html#cb178-19"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb178-20"><a href="variant-calling.html#cb178-20"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="variant-calling.html#cb179-1"></a><span class="kw">dir_info</span>(<span class="st">&quot;apps/variants/mutect2&quot;</span>, <span class="dt">regexp =</span> <span class="st">&quot;normal&quot;</span>, <span class="dt">recurse =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(path, size)</span></code></pre></div>
<pre><code>## # A tibble: 12 x 2
##    path                                                                size
##    &lt;fs::path&gt;                                                    &lt;fs::byte&gt;
##  1 apps/variants/mutect2/normal19                                        4K
##  2 apps/variants/mutect2/normal19/normal19.vcf                       15.96M
##  3 apps/variants/mutect2/normal19/normal19.vcf.idx                  117.92K
##  4 apps/variants/mutect2/normal19/normal19.vcf.stats                     37
##  5 apps/variants/mutect2/normal20                                        4K
##  6 apps/variants/mutect2/normal20/normal20.vcf                       14.59M
##  7 apps/variants/mutect2/normal20/normal20.vcf.idx                  248.43K
##  8 apps/variants/mutect2/normal20/normal20.vcf.stats                     37
##  9 apps/variants/mutect2/pt19/normal19.tumor19.ctfiltered.obfil…    542.26K
## 10 apps/variants/mutect2/pt19/normal19.tumor19.ctfiltered.obfil…      1.04M
## 11 apps/variants/mutect2/pt20/normal20.tumor20.ctfiltered.obfil…    517.15K
## 12 apps/variants/mutect2/pt20/normal20.tumor20.ctfiltered.obfil…    990.28K</code></pre>
<p>Note that, we need to add “-max-mnp-distance 0” to the arguments according to the Mutect2 documents to avoid a bug in “GenomicsDBImport” step.</p>
<ol start="2" style="list-style-type: decimal">
<li>Create a panel of normals (PoN)</li>
</ol>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="variant-calling.html#cb181-1"></a>GPoN<span class="op">$</span>nvcf &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">list.files</span>(<span class="st">&quot;apps/variants/mutect2/&quot;</span>, <span class="st">&quot;.vcf$&quot;</span>,</span>
<span id="cb181-2"><a href="variant-calling.html#cb181-2"></a>                                <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>))</span>
<span id="cb181-3"><a href="variant-calling.html#cb181-3"></a>GPoN<span class="op">$</span>Ref &lt;-<span class="st"> &quot;apps/data/hs37d5.fa&quot;</span></span>
<span id="cb181-4"><a href="variant-calling.html#cb181-4"></a>GPoN<span class="op">$</span>intervals &lt;-<span class="st"> &quot;apps/variants/vcf/intval.bed&quot;</span></span>
<span id="cb181-5"><a href="variant-calling.html#cb181-5"></a>GPoN<span class="op">$</span>pvcf &lt;-<span class="st"> &quot;pon.vcf&quot;</span></span>
<span id="cb181-6"><a href="variant-calling.html#cb181-6"></a><span class="co">##GPoN$gresource &lt;- &quot;apps/data/Mutect2/af-only-gnomad.raw.sites.b37.vcf&quot;</span></span>
<span id="cb181-7"><a href="variant-calling.html#cb181-7"></a><span class="kw">runCWL</span>(GPoN, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/mutect2/&quot;</span>, <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>, <span class="dt">Args =</span> <span class="st">&quot;--relax-path-checks&quot;</span>)</span></code></pre></div>
<p>Here we add “–relax-path-checks” arugment to the “cwltool” because the temporary files have special characters in their file names.</p>
<ol start="3" style="list-style-type: decimal">
<li>Mutect2 pipeline</li>
</ol>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="variant-calling.html#cb182-1"></a>tbam &lt;-<span class="st"> </span><span class="kw">as.list</span>(bams[<span class="dv">3</span><span class="op">:</span><span class="dv">4</span>])</span>
<span id="cb182-2"><a href="variant-calling.html#cb182-2"></a>nbam &lt;-<span class="st"> </span><span class="kw">as.list</span>(bams[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</span>
<span id="cb182-3"><a href="variant-calling.html#cb182-3"></a>nid &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">sub</span>(<span class="st">&quot;.bam&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">basename</span>(bams))[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</span>
<span id="cb182-4"><a href="variant-calling.html#cb182-4"></a>tid &lt;-<span class="st"> </span><span class="kw">as.list</span>(<span class="kw">sub</span>(<span class="st">&quot;.bam&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">basename</span>(bams))[<span class="dv">3</span><span class="op">:</span><span class="dv">4</span>])</span>
<span id="cb182-5"><a href="variant-calling.html#cb182-5"></a></span>
<span id="cb182-6"><a href="variant-calling.html#cb182-6"></a><span class="kw">names</span>(tbam) &lt;-<span class="st"> </span><span class="kw">names</span>(nbam) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;pt19&quot;</span>, <span class="st">&quot;pt20&quot;</span>)</span>
<span id="cb182-7"><a href="variant-calling.html#cb182-7"></a>inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">tbam =</span> tbam, <span class="dt">nbam =</span> nbam,</span>
<span id="cb182-8"><a href="variant-calling.html#cb182-8"></a>                  <span class="dt">normal =</span> nid, <span class="dt">tumor =</span> tid)</span>
<span id="cb182-9"><a href="variant-calling.html#cb182-9"></a>paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">Ref =</span> <span class="st">&quot;apps/data/hs37d5.fa&quot;</span>,</span>
<span id="cb182-10"><a href="variant-calling.html#cb182-10"></a>                  <span class="dt">gresource =</span> <span class="st">&quot;apps/data/Mutect2/af-only-gnomad.raw.sites.b37.vcf&quot;</span>,</span>
<span id="cb182-11"><a href="variant-calling.html#cb182-11"></a>                  <span class="dt">pon =</span> <span class="st">&quot;apps/variants/mutect2/pon.vcf&quot;</span>,</span>
<span id="cb182-12"><a href="variant-calling.html#cb182-12"></a>                  <span class="dt">interval =</span> <span class="st">&quot;apps/variants/vcf/intval.bed&quot;</span>,</span>
<span id="cb182-13"><a href="variant-calling.html#cb182-13"></a>                  <span class="dt">comvcf =</span> <span class="st">&quot;apps/data/Mutect2/GetPileupSummaries/small_exac_common_3_b37.vcf&quot;</span>)</span>
<span id="cb182-14"><a href="variant-calling.html#cb182-14"></a></span>
<span id="cb182-15"><a href="variant-calling.html#cb182-15"></a>res6 &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(Mutect2PL, <span class="dt">outdir =</span> <span class="st">&quot;apps/variants/mutect2&quot;</span>, inputList, paramList,</span>
<span id="cb182-16"><a href="variant-calling.html#cb182-16"></a>                    <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>,</span>
<span id="cb182-17"><a href="variant-calling.html#cb182-17"></a>                                              <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,</span>
<span id="cb182-18"><a href="variant-calling.html#cb182-18"></a>                                              <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,</span>
<span id="cb182-19"><a href="variant-calling.html#cb182-19"></a>                                              <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">16</span>,</span>
<span id="cb182-20"><a href="variant-calling.html#cb182-20"></a>                                                               <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>,</span>
<span id="cb182-21"><a href="variant-calling.html#cb182-21"></a>                                                               <span class="dt">jobname =</span> <span class="st">&quot;mutect2&quot;</span>),</span>
<span id="cb182-22"><a href="variant-calling.html#cb182-22"></a>                                              <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">logdir =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>),</span>
<span id="cb182-23"><a href="variant-calling.html#cb182-23"></a>                    <span class="dt">stderr =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
</div>
<div id="muse" class="section level3">
<h3><span class="header-section-number">6.3.2</span> MuSE</h3>
</div>
<div id="strelka2" class="section level3">
<h3><span class="header-section-number">6.3.3</span> Strelka2</h3>
</div>
<div id="somaticsniper" class="section level3">
<h3><span class="header-section-number">6.3.4</span> SomaticSniper</h3>
</div>
<div id="vardict" class="section level3">
<h3><span class="header-section-number">6.3.5</span> VarDict</h3>
</div>
<div id="lofreq" class="section level3">
<h3><span class="header-section-number">6.3.6</span> LoFreq</h3>
</div>
<div id="varscan2" class="section level3">
<h3><span class="header-section-number">6.3.7</span> VarScan2</h3>
</div>
<div id="lancet" class="section level3">
<h3><span class="header-section-number">6.3.8</span> Lancet</h3>
</div>
<div id="neusomatic" class="section level3">
<h3><span class="header-section-number">6.3.9</span> Neusomatic</h3>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="application.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="rna-seq-alignment-and-quantification.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/06-VariantCalling.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["Rcwl-Pipelines.pdf", "Rcwl-Pipelines.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
