<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 10 DNASeq alignment pipeline | Bioinformatics tools and pipelines with R and CWL</title>
  <meta name="description" content="A manual to wrap Bioinformatics tools and pipeplines with Common Workflow Language using the Rcwl package.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 10 DNASeq alignment pipeline | Bioinformatics tools and pipelines with R and CWL" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A manual to wrap Bioinformatics tools and pipeplines with Common Workflow Language using the Rcwl package." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 DNASeq alignment pipeline | Bioinformatics tools and pipelines with R and CWL" />
  
  <meta name="twitter:description" content="A manual to wrap Bioinformatics tools and pipeplines with Common Workflow Language using the Rcwl package." />
  

<meta name="author" content="Qiang Hu">


<meta name="date" content="2019-03-13">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="rcwlpipelines-tools.html">
<link rel="next" href="rnaseq-pipeline.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/d3-3.3.8/d3.min.js"></script>
<script src="libs/dagre-0.4.0/dagre-d3.min.js"></script>
<link href="libs/mermaid-0.3.0/dist/mermaid.css" rel="stylesheet" />
<script src="libs/mermaid-0.3.0/dist/mermaid.slim.min.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/chromatography-0.1/chromatography.js"></script>
<script src="libs/DiagrammeR-binding-1.0.0/DiagrammeR.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#r-package-installation"><i class="fa fa-check"></i><b>1.1</b> R package installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#system-requirements"><i class="fa fa-check"></i><b>1.2</b> System requirements</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#docker"><i class="fa fa-check"></i><b>1.3</b> Docker</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#structure-of-the-book"><i class="fa fa-check"></i><b>1.4</b> Structure of the book</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#r-session-information"><i class="fa fa-check"></i><b>1.5</b> R session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#common-workflow-language"><i class="fa fa-check"></i><b>2.1</b> Common Workflow Language</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#first-example"><i class="fa fa-check"></i><b>2.2</b> First example</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#test-run"><i class="fa fa-check"></i><b>2.3</b> Test run</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="input-parameters.html"><a href="input-parameters.html"><i class="fa fa-check"></i><b>3</b> Input Parameters</a><ul>
<li class="chapter" data-level="3.1" data-path="input-parameters.html"><a href="input-parameters.html#essential-input-parameters"><i class="fa fa-check"></i><b>3.1</b> Essential Input parameters</a></li>
<li class="chapter" data-level="3.2" data-path="input-parameters.html"><a href="input-parameters.html#array-inputs"><i class="fa fa-check"></i><b>3.2</b> Array Inputs</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="output-parameters.html"><a href="output-parameters.html"><i class="fa fa-check"></i><b>4</b> Output Parameters</a><ul>
<li class="chapter" data-level="4.1" data-path="output-parameters.html"><a href="output-parameters.html#capturing-output"><i class="fa fa-check"></i><b>4.1</b> Capturing Output</a></li>
<li class="chapter" data-level="4.2" data-path="output-parameters.html"><a href="output-parameters.html#array-outputs"><i class="fa fa-check"></i><b>4.2</b> Array Outputs</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="writing-pipeline.html"><a href="writing-pipeline.html"><i class="fa fa-check"></i><b>5</b> Writing Pipeline</a><ul>
<li class="chapter" data-level="5.1" data-path="writing-pipeline.html"><a href="writing-pipeline.html#scattering-pipeline"><i class="fa fa-check"></i><b>5.1</b> Scattering pipeline</a></li>
<li class="chapter" data-level="5.2" data-path="writing-pipeline.html"><a href="writing-pipeline.html#pipeline-plot"><i class="fa fa-check"></i><b>5.2</b> Pipeline plot</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="running-tools-in-docker.html"><a href="running-tools-in-docker.html"><i class="fa fa-check"></i><b>6</b> Running Tools in Docker</a></li>
<li class="chapter" data-level="7" data-path="running-tools-in-cluster-server.html"><a href="running-tools-in-cluster-server.html"><i class="fa fa-check"></i><b>7</b> Running Tools in Cluster server</a></li>
<li class="chapter" data-level="8" data-path="web-application.html"><a href="web-application.html"><i class="fa fa-check"></i><b>8</b> Web Application</a><ul>
<li class="chapter" data-level="8.1" data-path="web-application.html"><a href="web-application.html#cwlparam-example"><i class="fa fa-check"></i><b>8.1</b> cwlParam example</a></li>
<li class="chapter" data-level="8.2" data-path="web-application.html"><a href="web-application.html#cwlparam-to-shiny-app"><i class="fa fa-check"></i><b>8.2</b> cwlParam to Shiny App</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html"><i class="fa fa-check"></i><b>9</b> RcwlPipelines tools</a><ul>
<li class="chapter" data-level="9.1" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html#rcwl-scripts"><i class="fa fa-check"></i><b>9.1</b> Rcwl scripts</a></li>
<li class="chapter" data-level="9.2" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html#build-a-pipeline"><i class="fa fa-check"></i><b>9.2</b> Build a pipeline</a></li>
<li class="chapter" data-level="9.3" data-path="rcwlpipelines-tools.html"><a href="rcwlpipelines-tools.html#rcwlpipelines-summary"><i class="fa fa-check"></i><b>9.3</b> RcwlPipelines summary</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html"><i class="fa fa-check"></i><b>10</b> DNASeq alignment pipeline</a><ul>
<li class="chapter" data-level="10.1" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html#summary"><i class="fa fa-check"></i><b>10.1</b> Summary</a></li>
<li class="chapter" data-level="10.2" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html#prepara-data"><i class="fa fa-check"></i><b>10.2</b> Prepara data</a></li>
<li class="chapter" data-level="10.3" data-path="dnaseq-alignment-pipeline.html"><a href="dnaseq-alignment-pipeline.html#run-in-cluster"><i class="fa fa-check"></i><b>10.3</b> Run in cluster</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="rnaseq-pipeline.html"><a href="rnaseq-pipeline.html"><i class="fa fa-check"></i><b>11</b> RNASeq pipeline</a><ul>
<li class="chapter" data-level="11.1" data-path="rnaseq-pipeline.html"><a href="rnaseq-pipeline.html#prepare-data"><i class="fa fa-check"></i><b>11.1</b> Prepare data</a></li>
<li class="chapter" data-level="11.2" data-path="rnaseq-pipeline.html"><a href="rnaseq-pipeline.html#submit-pipeline-with-sge"><i class="fa fa-check"></i><b>11.2</b> Submit pipeline with SGE</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="summarize-qc.html"><a href="summarize-qc.html"><i class="fa fa-check"></i><b>12</b> Summarize QC</a></li>
<li class="chapter" data-level="13" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html"><i class="fa fa-check"></i><b>13</b> MC3 somatic variant calling pipeline</a><ul>
<li class="chapter" data-level="13.1" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#prepare-data-1"><i class="fa fa-check"></i><b>13.1</b> Prepare data</a></li>
<li class="chapter" data-level="13.2" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#run-mc3-pipeline"><i class="fa fa-check"></i><b>13.2</b> Run MC3 pipeline</a></li>
<li class="chapter" data-level="13.3" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#gatk4-germline-variant-calling-pipeline"><i class="fa fa-check"></i><b>13.3</b> GATK4 germline variant calling pipeline</a></li>
<li class="chapter" data-level="13.4" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#gatk-alignment"><i class="fa fa-check"></i><b>13.4</b> GATK Alignment</a></li>
<li class="chapter" data-level="13.5" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#haplotypecaller"><i class="fa fa-check"></i><b>13.5</b> HaplotypeCaller</a></li>
<li class="chapter" data-level="13.6" data-path="mc3-somatic-variant-calling-pipeline.html"><a href="mc3-somatic-variant-calling-pipeline.html#joint-discovery"><i class="fa fa-check"></i><b>13.6</b> Joint Discovery</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics tools and pipelines with R and CWL</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dnaseq-alignment-pipeline" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> DNASeq alignment pipeline</h1>
<div id="summary" class="section level2">
<h2><span class="header-section-number">10.1</span> Summary</h2>
<p>The pipeline can be used to preprocess DNA sequences in fastq
format. It can take paired fastqs, read groups from multiple batches
as input.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(alignMerge)
<span class="kw">inputs</span>(alignMerge)</code></pre>
<pre><code>## List of length 6
## names(6): idBam RG threads Ref FQ1s FQ2s</code></pre>
<p>The pipeline includes two steps and several jobs will be run in each
step.</p>
<ol style="list-style-type: decimal">
<li><code>bwaAlign</code>: bwa alignment by read groups.</li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">runs</span>(<span class="kw">runs</span>(alignMerge)[[<span class="dv">1</span>]])</code></pre>
<pre><code>## List of length 4
## names(4): bwa sam2bam sortBam idxBam</code></pre>
<ul>
<li><code>bwa</code>: To align fastqs and read groups to reference genome with <code>bwa</code>.</li>
<li><code>sam2bam</code>: To convert the alignments in “sam” format to “bam”
format with <code>samtools</code>.</li>
<li><code>sortBam</code>: To sort the “bam” file by coordinates with <code>samtools</code>.</li>
<li><code>idxBam</code>: To index “bam” file with <code>samtools</code>.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><code>mergeBamDup</code>: Merge by samples and markduplicates.</li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">runs</span>(<span class="kw">runs</span>(alignMerge)[[<span class="dv">2</span>]])</code></pre>
<pre><code>## List of length 4
## names(4): mergeBam markdup samtools_index samtools_flagstat</code></pre>
<ul>
<li><code>mergeBam</code>: To merge bam files from multiple batches with <code>picard</code>.</li>
<li><code>markdup</code>: To mark duplicates with <code>picard</code>.</li>
<li><code>samtools_index</code>: To index bam file with <code>samtools</code>.</li>
<li><code>samtools_flagstat</code>: To summarize flags in bam with <code>samtools</code>.</li>
</ul>
<p>The final bam files after duplicates marked, bam index, duplicates
matrix, and flag statistics summary will be in the output folder.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">outputs</span>(alignMerge)</code></pre>
<pre><code>## List of length 4
## names(4): oBam matrix Idx stat</code></pre>
<p>Here is the short summary and steps plot.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">short</span>(alignMerge)</code></pre>
<pre><code>## inputs:
## - idBam
## - RG
## - threads
## - Ref
## - FQ1s
## - FQ2s
## outputs:
## - oBam
## - matrix
## - Idx
## - stat
## steps:
## - bwaAlign
## - mergeBamDup</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotCWL</span>(alignMerge)</code></pre>
<div id="htmlwidget-cea13da135eec2af4897" style="width:672px;height:480px;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-cea13da135eec2af4897">{"x":{"diagram":"graph TB\nC(threads)-->K{bwaAlign}\nB(RG)-->K{bwaAlign}\nD(Ref)-->K{bwaAlign}\nE(FQ1s)-->K{bwaAlign}\nF(FQ2s)-->K{bwaAlign}\nK{bwaAlign}-->|Bam|L{mergeBamDup}\nA(idBam)-->L{mergeBamDup}\nL{mergeBamDup}-->G((oBam))\nL{mergeBamDup}-->H((matrix))\nL{mergeBamDup}-->I((Idx))\nL{mergeBamDup}-->J((stat))"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="prepara-data" class="section level2">
<h2><span class="header-section-number">10.2</span> Prepara data</h2>
<p>Here is an example of two samples. The “sample1” have two lanes of sequences and the “sample2” only has one pair of reads. The lists of reads1 <code>fq1</code>, reads2 <code>fq2</code>, read groups and output BAM names are defined in the <code>inputList</code>. The reference genome and number of threads to run the job are defined in the shared options, <code>paramList</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">fq1 &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sample1 =</span> <span class="kw">list</span>(<span class="st">&quot;apps/DNASeq/data/fq1_1.fq&quot;</span>,
                           <span class="st">&quot;apps/DNASeq/data/fq2_1.fq&quot;</span>),
            <span class="dt">sample2 =</span> <span class="kw">list</span>(<span class="st">&quot;apps/DNASeq/data/fq1_1.fq&quot;</span>))
fq2 &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sample1 =</span> <span class="kw">list</span>(<span class="st">&quot;apps/DNASeq/data/fq1_2.fq&quot;</span>,
                           <span class="st">&quot;apps/DNASeq/data/fq2_2.fq&quot;</span>),
            <span class="dt">sample2 =</span> <span class="kw">list</span>(<span class="st">&quot;apps/DNASeq/data/fq1_2.fq&quot;</span>))
rgs &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sample1 =</span> <span class="kw">list</span>(<span class="st">&quot;@RG</span><span class="ch">\\</span><span class="st">tID:sample1.1</span><span class="ch">\\</span><span class="st">tSM:sample1&quot;</span>,
                           <span class="st">&quot;@RG</span><span class="ch">\\</span><span class="st">tID:sample1.2</span><span class="ch">\\</span><span class="st">tSM:sample1&quot;</span>),
            <span class="dt">sample2 =</span> <span class="kw">list</span>(<span class="st">&quot;@RG</span><span class="ch">\\</span><span class="st">tID:sample2.1</span><span class="ch">\\</span><span class="st">tSM:sample2&quot;</span>))
samples &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">sample1 =</span> <span class="st">&quot;sample1.bam&quot;</span>, <span class="dt">sample2 =</span> <span class="st">&quot;sample2.bam&quot;</span>)

inputList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">idBam =</span> samples, <span class="dt">RG=</span> rgs, <span class="dt">FQ1s =</span> fq1, <span class="dt">FQ2s =</span> fq2) 
paramList &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">threads =</span> <span class="dv">2</span>,
                  <span class="dt">Ref =</span> <span class="st">&quot;apps/data/hs37d5.fa&quot;</span>)</code></pre>
</div>
<div id="run-in-cluster" class="section level2">
<h2><span class="header-section-number">10.3</span> Run in cluster</h2>
<pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">runCWLBatch</span>(alignMerge, <span class="dt">outdir =</span> <span class="st">&quot;output/BAM&quot;</span>,
                   <span class="dt">inputList =</span> inputList, <span class="dt">paramList =</span> paramList,
                   <span class="dt">BPPARAM =</span> <span class="kw">BatchtoolsParam</span>(<span class="dt">workers =</span> <span class="dv">2</span>, <span class="dt">cluster =</span> <span class="st">&quot;sge&quot;</span>,
                                             <span class="dt">template =</span> <span class="st">&quot;apps/data/sge.tmpl&quot;</span>,
                                             <span class="dt">resources =</span> <span class="kw">list</span>(<span class="dt">jobname=</span><span class="st">&quot;bwa&quot;</span>,
                                                              <span class="dt">threads =</span> <span class="dv">2</span>,
                                                              <span class="dt">queue =</span> <span class="st">&quot;all.q&quot;</span>)))</code></pre>
<p>List outputs:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dir</span>(<span class="st">&quot;apps/DNASeq/output/BAM/sample1&quot;</span>)</code></pre>
<pre><code>## [1] &quot;sample1.bam&quot;          &quot;sample1.bam.bai&quot;      &quot;sample1.flagstat.txt&quot;
## [4] &quot;sample1.markdup.txt&quot;</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="rcwlpipelines-tools.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="rnaseq-pipeline.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/05-application.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
